{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "CoLab_TermProj.ipynb의 사본",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "_SmTogjeMD47"
      },
      "source": [
        "# 생물정보학 및 실습 1 - Term Project용 환경 셋업\n",
        "\n",
        "```\n",
        "생물정보학 및 실습 1\n",
        "서울대학교 협동과정 생물정보학전공 2021년 1학기\n",
        "\n",
        "이 노트북은 Creative Commons BY-SA 조건으로 재사용 가능합니다.\n",
        "장혜식 <hyeshik@snu.ac.kr>, 서울대학교 생명과학부\n",
        "```\n",
        "\n",
        "이 노트북은 Google Colab을 이용해서 Term Project를 할 수 있도록 환경을 준비해 줍니다. 터미널 사용이 원활하지 않은 상태이고, 데이터를 오랫동안 보관하기 어렵기 때문에 되도록 개인용 컴퓨터에서 [Anaconda](https://www.anaconda.com/)나 [JupyterLab](https://jupyterlab.readthedocs.io/en/stable/) 등을 설치하고 사용하시는 것이 편하지만, 개인용 컴퓨터 사양이 부족한 경우에는 Google CoLab을 사용할 수도 있습니다. 개인용 컴퓨터를 쓸 때는 Windows에서는 [WSL2](https://docs.microsoft.com/en-us/windows/wsl/install-win10)를 쓰시면 리눅스 커맨드라인 환경을 쓸 수 있습니다. Google Colab은 창을 닫은 후 90분 후, 또는 창을 처음 열고 12시간 후에 환경을 닫아버리기 때문에 매번 실습할 때마다 새로 셋업이 필요합니다. `Runtime->Run All`을 실행해서 전체를 한 번 실행하시면 준비됩니다. 모든 것이 잘 안 되고 뭔가 꼬인 느낌이 강하게 들 때는 `Runtime->Factory reset runtime`을 누르고 새로 시작하시면 될 수도 있습니다.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "zGYjheLcRJS0"
      },
      "source": [
        "## Google Drive 마운트\n",
        "\n",
        "데이터 파일들을 불러오고 저장하기 위해서 Google Drive를 연결합니다. **이번엔 스크립트 파일을 편집해야 하기 때문에 Google Drive 마운트를 하지 않으면 vim등 콘솔에서 쓸 수 있는 에디터를 쓰셔야 합니다.** 화면에 나오는 주소를 눌러서 인증하고 인증 코드를 그 아래에 입력하세요. 그러면 `/contents/drive/MyDrive`에서 자신의 Drive 최상단 폴더가 보입니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "0HgOxANiRHgl",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "75ffacdc-e886-4efe-b02c-e94c145fd12e"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/drive/')"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Mounted at /content/drive/\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ra3RB22DWEER"
      },
      "source": [
        "## 실습용 Bioconda 환경 설치하기\n",
        "실습용 프로그램들을 쉽게 설치할 수 있도록 [Bioconda](https://bioconda.github.io/) 환경을 만듭니다. 실행하고 진행상황을 관찰하다가 에러로 보이는 것이 있으면 잘 읽어보고 해결합니다.\n",
        "\n",
        "파이썬 모듈들은 Colab쪽의 파이썬 인터프리터를 바꿀 수 없기 때문에 Conda에 설치하더라도 사용하기 어렵습니다. 파이썬 모듈은 `pip` 명령어로 설치하시면 Colab에서 불러와서 사용할 수 있습니다.\n",
        "\n",
        "환경 초기화에 관련된 스크립트는 [GitHub 프로젝트](https://github.com/hyeshik/colab-biolab)에 있습니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "TTXmNi4hcap3",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "9c213449-42e1-47e8-fb99-28a31fbb9e11"
      },
      "source": [
        "!git clone https://github.com/hyeshik/colab-biolab.git\n",
        "!cd colab-biolab && bash tools/setup.sh\n",
        "exec(open('colab-biolab/tools/activate_conda.py').read())"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'colab-biolab'...\n",
            "remote: Enumerating objects: 69, done.\u001b[K\n",
            "remote: Counting objects: 100% (69/69), done.\u001b[K\n",
            "remote: Compressing objects: 100% (43/43), done.\u001b[K\n",
            "remote: Total 69 (delta 22), reused 54 (delta 13), pack-reused 0\u001b[K\n",
            "Unpacking objects: 100% (69/69), done.\n",
            "./\n",
            "./root/\n",
            "./root/.vimrc\n",
            "./root/.condarc\n",
            "./root/.tmux.conf\n",
            "./root/.bashrc.biolab\n",
            "./root/.profile\n",
            "./root/.bin.priority/\n",
            "./root/.bin.priority/pip\n",
            "./root/.bin.priority/pip2\n",
            "./root/.bin.priority/pip3\n",
            "--2021-05-11 10:15:26--  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
            "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8203, ...\n",
            "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 94235922 (90M) [application/x-sh]\n",
            "Saving to: ‘miniconda3.sh’\n",
            "\n",
            "miniconda3.sh       100%[===================>]  89.87M   168MB/s    in 0.5s    \n",
            "\n",
            "2021-05-11 10:15:26 (168 MB/s) - ‘miniconda3.sh’ saved [94235922/94235922]\n",
            "\n",
            "PREFIX=/root/conda\n",
            "Unpacking payload ...\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\bdone\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda\n",
            "\n",
            "  added / updated specs:\n",
            "    - _libgcc_mutex==0.1=main\n",
            "    - brotlipy==0.7.0=py38h27cfd23_1003\n",
            "    - ca-certificates==2020.10.14=0\n",
            "    - certifi==2020.6.20=pyhd3eb1b0_3\n",
            "    - cffi==1.14.3=py38h261ae71_2\n",
            "    - chardet==3.0.4=py38h06a4308_1003\n",
            "    - conda-package-handling==1.7.2=py38h03888b9_0\n",
            "    - conda==4.9.2=py38h06a4308_0\n",
            "    - cryptography==3.2.1=py38h3c74f83_1\n",
            "    - idna==2.10=py_0\n",
            "    - ld_impl_linux-64==2.33.1=h53a641e_7\n",
            "    - libedit==3.1.20191231=h14c3975_1\n",
            "    - libffi==3.3=he6710b0_2\n",
            "    - libgcc-ng==9.1.0=hdf63c60_0\n",
            "    - libstdcxx-ng==9.1.0=hdf63c60_0\n",
            "    - ncurses==6.2=he6710b0_1\n",
            "    - openssl==1.1.1h=h7b6447c_0\n",
            "    - pip==20.2.4=py38h06a4308_0\n",
            "    - pycosat==0.6.3=py38h7b6447c_1\n",
            "    - pycparser==2.20=py_2\n",
            "    - pyopenssl==19.1.0=pyhd3eb1b0_1\n",
            "    - pysocks==1.7.1=py38h06a4308_0\n",
            "    - python==3.8.5=h7579374_1\n",
            "    - readline==8.0=h7b6447c_0\n",
            "    - requests==2.24.0=py_0\n",
            "    - ruamel_yaml==0.15.87=py38h7b6447c_1\n",
            "    - setuptools==50.3.1=py38h06a4308_1\n",
            "    - six==1.15.0=py38h06a4308_0\n",
            "    - sqlite==3.33.0=h62c20be_0\n",
            "    - tk==8.6.10=hbc83047_0\n",
            "    - tqdm==4.51.0=pyhd3eb1b0_0\n",
            "    - urllib3==1.25.11=py_0\n",
            "    - wheel==0.35.1=pyhd3eb1b0_0\n",
            "    - xz==5.2.5=h7b6447c_0\n",
            "    - yaml==0.2.5=h7b6447c_0\n",
            "    - zlib==1.2.11=h7b6447c_3\n",
            "\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
            "  brotlipy           pkgs/main/linux-64::brotlipy-0.7.0-py38h27cfd23_1003\n",
            "  ca-certificates    pkgs/main/linux-64::ca-certificates-2020.10.14-0\n",
            "  certifi            pkgs/main/noarch::certifi-2020.6.20-pyhd3eb1b0_3\n",
            "  cffi               pkgs/main/linux-64::cffi-1.14.3-py38h261ae71_2\n",
            "  chardet            pkgs/main/linux-64::chardet-3.0.4-py38h06a4308_1003\n",
            "  conda              pkgs/main/linux-64::conda-4.9.2-py38h06a4308_0\n",
            "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.7.2-py38h03888b9_0\n",
            "  cryptography       pkgs/main/linux-64::cryptography-3.2.1-py38h3c74f83_1\n",
            "  idna               pkgs/main/noarch::idna-2.10-py_0\n",
            "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7\n",
            "  libedit            pkgs/main/linux-64::libedit-3.1.20191231-h14c3975_1\n",
            "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2\n",
            "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0\n",
            "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0\n",
            "  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1\n",
            "  openssl            pkgs/main/linux-64::openssl-1.1.1h-h7b6447c_0\n",
            "  pip                pkgs/main/linux-64::pip-20.2.4-py38h06a4308_0\n",
            "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py38h7b6447c_1\n",
            "  pycparser          pkgs/main/noarch::pycparser-2.20-py_2\n",
            "  pyopenssl          pkgs/main/noarch::pyopenssl-19.1.0-pyhd3eb1b0_1\n",
            "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py38h06a4308_0\n",
            "  python             pkgs/main/linux-64::python-3.8.5-h7579374_1\n",
            "  readline           pkgs/main/linux-64::readline-8.0-h7b6447c_0\n",
            "  requests           pkgs/main/noarch::requests-2.24.0-py_0\n",
            "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.87-py38h7b6447c_1\n",
            "  setuptools         pkgs/main/linux-64::setuptools-50.3.1-py38h06a4308_1\n",
            "  six                pkgs/main/linux-64::six-1.15.0-py38h06a4308_0\n",
            "  sqlite             pkgs/main/linux-64::sqlite-3.33.0-h62c20be_0\n",
            "  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0\n",
            "  tqdm               pkgs/main/noarch::tqdm-4.51.0-pyhd3eb1b0_0\n",
            "  urllib3            pkgs/main/noarch::urllib3-1.25.11-py_0\n",
            "  wheel              pkgs/main/noarch::wheel-0.35.1-pyhd3eb1b0_0\n",
            "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
            "  yaml               pkgs/main/linux-64::yaml-0.2.5-h7b6447c_0\n",
            "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
            "\n",
            "\n",
            "Preparing transaction: / \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Executing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "installation finished.\n",
            "WARNING:\n",
            "    You currently have a PYTHONPATH environment variable set. This may cause\n",
            "    unexpected behavior when running the Python interpreter in Miniconda3.\n",
            "    For best results, please verify that your PYTHONPATH only points to\n",
            "    directories of packages that are compatible with the Python interpreter\n",
            "    in Miniconda3: /root/conda\n",
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "Solving environment: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda/envs/lab\n",
            "\n",
            "  added / updated specs:\n",
            "    - bedtools\n",
            "    - biopython\n",
            "    - bwa\n",
            "    - cython\n",
            "    - matplotlib\n",
            "    - numpy\n",
            "    - pandas\n",
            "    - pip\n",
            "    - pysam\n",
            "    - samtools\n",
            "    - scipy\n",
            "    - star\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    _libgcc_mutex-0.1          |      conda_forge           3 KB  conda-forge\n",
            "    _openmp_mutex-4.5          |            1_gnu          22 KB  conda-forge\n",
            "    bedtools-2.30.0            |       h7d7f7ad_1        17.9 MB  bioconda\n",
            "    biopython-1.78             |   py37h5e8e339_2         2.6 MB  conda-forge\n",
            "    bwa-0.7.17                 |       h5bf99c6_8         608 KB  bioconda\n",
            "    bzip2-1.0.8                |       h7f98852_4         484 KB  conda-forge\n",
            "    c-ares-1.17.1              |       h7f98852_1         109 KB  conda-forge\n",
            "    ca-certificates-2020.12.5  |       ha878542_0         137 KB  conda-forge\n",
            "    certifi-2020.12.5          |   py37h89c1867_1         143 KB  conda-forge\n",
            "    cycler-0.10.0              |             py_2           9 KB  conda-forge\n",
            "    cython-0.29.23             |   py37hcd2ae1e_0         2.2 MB  conda-forge\n",
            "    dbus-1.13.6                |       h48d8840_2         572 KB  conda-forge\n",
            "    expat-2.3.0                |       h9c3ff4c_0         168 KB  conda-forge\n",
            "    fontconfig-2.13.1          |    hba837de_1005         357 KB  conda-forge\n",
            "    freetype-2.10.4            |       h0708190_1         890 KB  conda-forge\n",
            "    gettext-0.19.8.1           |    h0b5b191_1005         3.6 MB  conda-forge\n",
            "    glib-2.68.1                |       h9c3ff4c_0         442 KB  conda-forge\n",
            "    glib-tools-2.68.1          |       h9c3ff4c_0          86 KB  conda-forge\n",
            "    gst-plugins-base-1.14.0    |       hbbd80ab_1         4.8 MB\n",
            "    gstreamer-1.14.0           |       h28cd5cc_2         3.2 MB\n",
            "    htslib-1.12                |       h9093b5e_1         2.1 MB  bioconda\n",
            "    icu-58.2                   |    hf484d3e_1000        22.6 MB  conda-forge\n",
            "    jpeg-9d                    |       h36c2ea0_0         264 KB  conda-forge\n",
            "    kiwisolver-1.3.1           |   py37h2527ec5_1          78 KB  conda-forge\n",
            "    krb5-1.17.2                |       h926e7f8_0         1.4 MB  conda-forge\n",
            "    lcms2-2.12                 |       hddcbb42_0         443 KB  conda-forge\n",
            "    ld_impl_linux-64-2.35.1    |       hea4e1c9_2         618 KB  conda-forge\n",
            "    libblas-3.9.0              |       9_openblas          11 KB  conda-forge\n",
            "    libcblas-3.9.0             |       9_openblas          11 KB  conda-forge\n",
            "    libcurl-7.76.1             |       hc4aaa36_1         328 KB  conda-forge\n",
            "    libdeflate-1.7             |       h7f98852_5          67 KB  conda-forge\n",
            "    libedit-3.1.20191231       |       he28a2e2_2         121 KB  conda-forge\n",
            "    libev-4.33                 |       h516909a_1         104 KB  conda-forge\n",
            "    libffi-3.3                 |       h58526e2_2          51 KB  conda-forge\n",
            "    libgcc-ng-9.3.0            |      h2828fa1_19         7.8 MB  conda-forge\n",
            "    libgfortran-ng-9.3.0       |      hff62375_19          22 KB  conda-forge\n",
            "    libgfortran5-9.3.0         |      hff62375_19         2.0 MB  conda-forge\n",
            "    libglib-2.68.1             |       h3e27bee_0         3.1 MB  conda-forge\n",
            "    libgomp-9.3.0              |      h2828fa1_19         376 KB  conda-forge\n",
            "    libiconv-1.16              |       h516909a_0         1.4 MB  conda-forge\n",
            "    liblapack-3.9.0            |       9_openblas          11 KB  conda-forge\n",
            "    libnghttp2-1.43.0          |       h812cca2_0         808 KB  conda-forge\n",
            "    libopenblas-0.3.15         |pthreads_h8fe5266_0         9.2 MB  conda-forge\n",
            "    libpng-1.6.37              |       h21135ba_2         306 KB  conda-forge\n",
            "    libssh2-1.9.0              |       ha56f1ee_6         226 KB  conda-forge\n",
            "    libstdcxx-ng-9.3.0         |      h6de172a_19         4.0 MB  conda-forge\n",
            "    libtiff-4.2.0              |       hdc55705_1         678 KB  conda-forge\n",
            "    libuuid-2.32.1             |    h7f98852_1000          28 KB  conda-forge\n",
            "    libwebp-base-1.2.0         |       h7f98852_2         815 KB  conda-forge\n",
            "    libxcb-1.13                |    h7f98852_1003         395 KB  conda-forge\n",
            "    libxml2-2.9.10             |       hb55368b_3         1.2 MB\n",
            "    lz4-c-1.9.3                |       h9c3ff4c_0         179 KB  conda-forge\n",
            "    matplotlib-3.4.2           |   py37h89c1867_0           7 KB  conda-forge\n",
            "    matplotlib-base-3.4.2      |   py37hdd32ed1_0         7.2 MB  conda-forge\n",
            "    ncurses-6.2                |       h58526e2_4         985 KB  conda-forge\n",
            "    numpy-1.20.2               |   py37h038b26d_0         5.8 MB  conda-forge\n",
            "    olefile-0.46               |     pyh9f0ad1d_1          32 KB  conda-forge\n",
            "    openjpeg-2.4.0             |       hf7af979_0         525 KB  conda-forge\n",
            "    openssl-1.1.1k             |       h7f98852_0         2.1 MB  conda-forge\n",
            "    pandas-1.2.4               |   py37h219a48f_0        11.8 MB  conda-forge\n",
            "    pcre-8.44                  |       he1b5a44_0         261 KB  conda-forge\n",
            "    perl-5.32.0                |       h36c2ea0_0        16.8 MB  conda-forge\n",
            "    pillow-8.1.2               |   py37h4600e1f_1         688 KB  conda-forge\n",
            "    pip-21.1.1                 |     pyhd8ed1ab_0         1.1 MB  conda-forge\n",
            "    pthread-stubs-0.4          |    h36c2ea0_1001           5 KB  conda-forge\n",
            "    pyparsing-2.4.7            |     pyh9f0ad1d_0          60 KB  conda-forge\n",
            "    pyqt-5.9.2                 |   py37hcca6a23_4         5.7 MB  conda-forge\n",
            "    pysam-0.16.0.1             |   py37h45aed0b_3         2.6 MB  bioconda\n",
            "    python-3.7.10              |hffdb5ce_100_cpython        57.3 MB  conda-forge\n",
            "    python-dateutil-2.8.1      |             py_0         220 KB  conda-forge\n",
            "    python_abi-3.7             |          1_cp37m           4 KB  conda-forge\n",
            "    pytz-2021.1                |     pyhd8ed1ab_0         239 KB  conda-forge\n",
            "    qt-5.9.7                   |       h5867ecd_1        68.5 MB\n",
            "    readline-8.1               |       h46c0cb4_0         295 KB  conda-forge\n",
            "    samtools-1.12              |       h9aed4be_1         386 KB  bioconda\n",
            "    scipy-1.6.3                |   py37h29e03ee_0        20.5 MB  conda-forge\n",
            "    setuptools-49.6.0          |   py37h89c1867_3         947 KB  conda-forge\n",
            "    sip-4.19.8                 |   py37hf484d3e_0         274 KB\n",
            "    six-1.16.0                 |     pyh6c4a22f_0          14 KB  conda-forge\n",
            "    sqlite-3.35.5              |       h74cdb3f_0         1.4 MB  conda-forge\n",
            "    star-2.7.8a                |       h9ee0642_1         3.8 MB  bioconda\n",
            "    tk-8.6.10                  |       h21135ba_1         3.2 MB  conda-forge\n",
            "    tornado-6.1                |   py37h5e8e339_1         646 KB  conda-forge\n",
            "    wheel-0.36.2               |     pyhd3deb0d_0          31 KB  conda-forge\n",
            "    xorg-libxau-1.0.9          |       h7f98852_0          13 KB  conda-forge\n",
            "    xorg-libxdmcp-1.1.3        |       h7f98852_0          19 KB  conda-forge\n",
            "    xz-5.2.5                   |       h516909a_1         343 KB  conda-forge\n",
            "    zlib-1.2.11                |    h516909a_1010         106 KB  conda-forge\n",
            "    zstd-1.4.9                 |       ha95c52a_0         431 KB  conda-forge\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:       312.9 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  _libgcc_mutex      conda-forge/linux-64::_libgcc_mutex-0.1-conda_forge\n",
            "  _openmp_mutex      conda-forge/linux-64::_openmp_mutex-4.5-1_gnu\n",
            "  bedtools           bioconda/linux-64::bedtools-2.30.0-h7d7f7ad_1\n",
            "  biopython          conda-forge/linux-64::biopython-1.78-py37h5e8e339_2\n",
            "  bwa                bioconda/linux-64::bwa-0.7.17-h5bf99c6_8\n",
            "  bzip2              conda-forge/linux-64::bzip2-1.0.8-h7f98852_4\n",
            "  c-ares             conda-forge/linux-64::c-ares-1.17.1-h7f98852_1\n",
            "  ca-certificates    conda-forge/linux-64::ca-certificates-2020.12.5-ha878542_0\n",
            "  certifi            conda-forge/linux-64::certifi-2020.12.5-py37h89c1867_1\n",
            "  cycler             conda-forge/noarch::cycler-0.10.0-py_2\n",
            "  cython             conda-forge/linux-64::cython-0.29.23-py37hcd2ae1e_0\n",
            "  dbus               conda-forge/linux-64::dbus-1.13.6-h48d8840_2\n",
            "  expat              conda-forge/linux-64::expat-2.3.0-h9c3ff4c_0\n",
            "  fontconfig         conda-forge/linux-64::fontconfig-2.13.1-hba837de_1005\n",
            "  freetype           conda-forge/linux-64::freetype-2.10.4-h0708190_1\n",
            "  gettext            conda-forge/linux-64::gettext-0.19.8.1-h0b5b191_1005\n",
            "  glib               conda-forge/linux-64::glib-2.68.1-h9c3ff4c_0\n",
            "  glib-tools         conda-forge/linux-64::glib-tools-2.68.1-h9c3ff4c_0\n",
            "  gst-plugins-base   pkgs/main/linux-64::gst-plugins-base-1.14.0-hbbd80ab_1\n",
            "  gstreamer          pkgs/main/linux-64::gstreamer-1.14.0-h28cd5cc_2\n",
            "  htslib             bioconda/linux-64::htslib-1.12-h9093b5e_1\n",
            "  icu                conda-forge/linux-64::icu-58.2-hf484d3e_1000\n",
            "  jpeg               conda-forge/linux-64::jpeg-9d-h36c2ea0_0\n",
            "  kiwisolver         conda-forge/linux-64::kiwisolver-1.3.1-py37h2527ec5_1\n",
            "  krb5               conda-forge/linux-64::krb5-1.17.2-h926e7f8_0\n",
            "  lcms2              conda-forge/linux-64::lcms2-2.12-hddcbb42_0\n",
            "  ld_impl_linux-64   conda-forge/linux-64::ld_impl_linux-64-2.35.1-hea4e1c9_2\n",
            "  libblas            conda-forge/linux-64::libblas-3.9.0-9_openblas\n",
            "  libcblas           conda-forge/linux-64::libcblas-3.9.0-9_openblas\n",
            "  libcurl            conda-forge/linux-64::libcurl-7.76.1-hc4aaa36_1\n",
            "  libdeflate         conda-forge/linux-64::libdeflate-1.7-h7f98852_5\n",
            "  libedit            conda-forge/linux-64::libedit-3.1.20191231-he28a2e2_2\n",
            "  libev              conda-forge/linux-64::libev-4.33-h516909a_1\n",
            "  libffi             conda-forge/linux-64::libffi-3.3-h58526e2_2\n",
            "  libgcc-ng          conda-forge/linux-64::libgcc-ng-9.3.0-h2828fa1_19\n",
            "  libgfortran-ng     conda-forge/linux-64::libgfortran-ng-9.3.0-hff62375_19\n",
            "  libgfortran5       conda-forge/linux-64::libgfortran5-9.3.0-hff62375_19\n",
            "  libglib            conda-forge/linux-64::libglib-2.68.1-h3e27bee_0\n",
            "  libgomp            conda-forge/linux-64::libgomp-9.3.0-h2828fa1_19\n",
            "  libiconv           conda-forge/linux-64::libiconv-1.16-h516909a_0\n",
            "  liblapack          conda-forge/linux-64::liblapack-3.9.0-9_openblas\n",
            "  libnghttp2         conda-forge/linux-64::libnghttp2-1.43.0-h812cca2_0\n",
            "  libopenblas        conda-forge/linux-64::libopenblas-0.3.15-pthreads_h8fe5266_0\n",
            "  libpng             conda-forge/linux-64::libpng-1.6.37-h21135ba_2\n",
            "  libssh2            conda-forge/linux-64::libssh2-1.9.0-ha56f1ee_6\n",
            "  libstdcxx-ng       conda-forge/linux-64::libstdcxx-ng-9.3.0-h6de172a_19\n",
            "  libtiff            conda-forge/linux-64::libtiff-4.2.0-hdc55705_1\n",
            "  libuuid            conda-forge/linux-64::libuuid-2.32.1-h7f98852_1000\n",
            "  libwebp-base       conda-forge/linux-64::libwebp-base-1.2.0-h7f98852_2\n",
            "  libxcb             conda-forge/linux-64::libxcb-1.13-h7f98852_1003\n",
            "  libxml2            pkgs/main/linux-64::libxml2-2.9.10-hb55368b_3\n",
            "  lz4-c              conda-forge/linux-64::lz4-c-1.9.3-h9c3ff4c_0\n",
            "  matplotlib         conda-forge/linux-64::matplotlib-3.4.2-py37h89c1867_0\n",
            "  matplotlib-base    conda-forge/linux-64::matplotlib-base-3.4.2-py37hdd32ed1_0\n",
            "  ncurses            conda-forge/linux-64::ncurses-6.2-h58526e2_4\n",
            "  numpy              conda-forge/linux-64::numpy-1.20.2-py37h038b26d_0\n",
            "  olefile            conda-forge/noarch::olefile-0.46-pyh9f0ad1d_1\n",
            "  openjpeg           conda-forge/linux-64::openjpeg-2.4.0-hf7af979_0\n",
            "  openssl            conda-forge/linux-64::openssl-1.1.1k-h7f98852_0\n",
            "  pandas             conda-forge/linux-64::pandas-1.2.4-py37h219a48f_0\n",
            "  pcre               conda-forge/linux-64::pcre-8.44-he1b5a44_0\n",
            "  perl               conda-forge/linux-64::perl-5.32.0-h36c2ea0_0\n",
            "  pillow             conda-forge/linux-64::pillow-8.1.2-py37h4600e1f_1\n",
            "  pip                conda-forge/noarch::pip-21.1.1-pyhd8ed1ab_0\n",
            "  pthread-stubs      conda-forge/linux-64::pthread-stubs-0.4-h36c2ea0_1001\n",
            "  pyparsing          conda-forge/noarch::pyparsing-2.4.7-pyh9f0ad1d_0\n",
            "  pyqt               conda-forge/linux-64::pyqt-5.9.2-py37hcca6a23_4\n",
            "  pysam              bioconda/linux-64::pysam-0.16.0.1-py37h45aed0b_3\n",
            "  python             conda-forge/linux-64::python-3.7.10-hffdb5ce_100_cpython\n",
            "  python-dateutil    conda-forge/noarch::python-dateutil-2.8.1-py_0\n",
            "  python_abi         conda-forge/linux-64::python_abi-3.7-1_cp37m\n",
            "  pytz               conda-forge/noarch::pytz-2021.1-pyhd8ed1ab_0\n",
            "  qt                 pkgs/main/linux-64::qt-5.9.7-h5867ecd_1\n",
            "  readline           conda-forge/linux-64::readline-8.1-h46c0cb4_0\n",
            "  samtools           bioconda/linux-64::samtools-1.12-h9aed4be_1\n",
            "  scipy              conda-forge/linux-64::scipy-1.6.3-py37h29e03ee_0\n",
            "  setuptools         conda-forge/linux-64::setuptools-49.6.0-py37h89c1867_3\n",
            "  sip                pkgs/main/linux-64::sip-4.19.8-py37hf484d3e_0\n",
            "  six                conda-forge/noarch::six-1.16.0-pyh6c4a22f_0\n",
            "  sqlite             conda-forge/linux-64::sqlite-3.35.5-h74cdb3f_0\n",
            "  star               bioconda/linux-64::star-2.7.8a-h9ee0642_1\n",
            "  tk                 conda-forge/linux-64::tk-8.6.10-h21135ba_1\n",
            "  tornado            conda-forge/linux-64::tornado-6.1-py37h5e8e339_1\n",
            "  wheel              conda-forge/noarch::wheel-0.36.2-pyhd3deb0d_0\n",
            "  xorg-libxau        conda-forge/linux-64::xorg-libxau-1.0.9-h7f98852_0\n",
            "  xorg-libxdmcp      conda-forge/linux-64::xorg-libxdmcp-1.1.3-h7f98852_0\n",
            "  xz                 conda-forge/linux-64::xz-5.2.5-h516909a_1\n",
            "  zlib               conda-forge/linux-64::zlib-1.2.11-h516909a_1010\n",
            "  zstd               conda-forge/linux-64::zstd-1.4.9-ha95c52a_0\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "ca-certificates-2020 | 137 KB    | : 100% 1.0/1 [00:00<00:00, 10.67it/s]\n",
            "jpeg-9d              | 264 KB    | : 100% 1.0/1 [00:00<00:00, 10.00it/s]\n",
            "gst-plugins-base-1.1 | 4.8 MB    | : 100% 1.0/1 [00:00<00:00,  3.12it/s]\n",
            "libgfortran5-9.3.0   | 2.0 MB    | : 100% 1.0/1 [00:00<00:00,  1.77it/s]\n",
            "pyparsing-2.4.7      | 60 KB     | : 100% 1.0/1 [00:00<00:00, 21.84it/s]\n",
            "pthread-stubs-0.4    | 5 KB      | : 100% 1.0/1 [00:00<00:00, 23.89it/s]\n",
            "glib-tools-2.68.1    | 86 KB     | : 100% 1.0/1 [00:00<00:00, 20.99it/s]\n",
            "htslib-1.12          | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.81it/s]\n",
            "liblapack-3.9.0      | 11 KB     | : 100% 1.0/1 [00:00<00:00, 25.86it/s]\n",
            "libuuid-2.32.1       | 28 KB     | : 100% 1.0/1 [00:00<00:00, 29.16it/s]\n",
            "libcurl-7.76.1       | 328 KB    | : 100% 1.0/1 [00:00<00:00,  9.68it/s]\n",
            "bwa-0.7.17           | 608 KB    | : 100% 1.0/1 [00:00<00:00,  6.52it/s]\n",
            "star-2.7.8a          | 3.8 MB    | : 100% 1.0/1 [00:00<00:00,  1.17it/s]\n",
            "python_abi-3.7       | 4 KB      | : 100% 1.0/1 [00:00<00:00, 35.89it/s]\n",
            "zlib-1.2.11          | 106 KB    | : 100% 1.0/1 [00:00<00:00, 17.86it/s]\n",
            "libxml2-2.9.10       | 1.2 MB    | : 100% 1.0/1 [00:00<00:00,  8.17it/s]\n",
            "libglib-2.68.1       | 3.1 MB    | : 100% 1.0/1 [00:00<00:00,  1.13it/s]\n",
            "libffi-3.3           | 51 KB     | : 100% 1.0/1 [00:00<00:00, 25.52it/s]\n",
            "libpng-1.6.37        | 306 KB    | : 100% 1.0/1 [00:00<00:00,  7.07it/s]\n",
            "_openmp_mutex-4.5    | 22 KB     | : 100% 1.0/1 [00:00<00:00, 23.27it/s]\n",
            "lz4-c-1.9.3          | 179 KB    | : 100% 1.0/1 [00:00<00:00, 11.65it/s]\n",
            "glib-2.68.1          | 442 KB    | : 100% 1.0/1 [00:00<00:00,  4.31it/s]\n",
            "tornado-6.1          | 646 KB    | : 100% 1.0/1 [00:00<00:00,  3.64it/s]\n",
            "pip-21.1.1           | 1.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.33it/s]\n",
            "libxcb-1.13          | 395 KB    | : 100% 1.0/1 [00:00<00:00,  5.04it/s]\n",
            "libopenblas-0.3.15   | 9.2 MB    | : 100% 1.0/1 [00:02<00:00,  2.27s/it]\n",
            "sip-4.19.8           | 274 KB    | : 100% 1.0/1 [00:00<00:00, 15.18it/s]\n",
            "libgfortran-ng-9.3.0 | 22 KB     | : 100% 1.0/1 [00:00<00:00, 29.57it/s]\n",
            "kiwisolver-1.3.1     | 78 KB     | : 100% 1.0/1 [00:00<00:00, 21.54it/s]\n",
            "olefile-0.46         | 32 KB     | : 100% 1.0/1 [00:00<00:00, 17.19it/s]\n",
            "certifi-2020.12.5    | 143 KB    | : 100% 1.0/1 [00:00<00:00, 22.60it/s]\n",
            "six-1.16.0           | 14 KB     | : 100% 1.0/1 [00:00<00:00, 33.39it/s]\n",
            "scipy-1.6.3          | 20.5 MB   | : 100% 1.0/1 [00:04<00:00,  4.81s/it]               \n",
            "bedtools-2.30.0      | 17.9 MB   | : 100% 1.0/1 [00:03<00:00,  3.95s/it]\n",
            "krb5-1.17.2          | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  2.64it/s]\n",
            "libcblas-3.9.0       | 11 KB     | : 100% 1.0/1 [00:00<00:00, 24.31it/s]\n",
            "ld_impl_linux-64-2.3 | 618 KB    | : 100% 1.0/1 [00:00<00:00,  5.12it/s]\n",
            "libiconv-1.16        | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  3.91it/s]\n",
            "libwebp-base-1.2.0   | 815 KB    | : 100% 1.0/1 [00:00<00:00,  4.47it/s]\n",
            "icu-58.2             | 22.6 MB   | : 100% 1.0/1 [00:04<00:00,  4.75s/it]\n",
            "libblas-3.9.0        | 11 KB     | : 100% 1.0/1 [00:00<00:00, 33.92it/s]\n",
            "libtiff-4.2.0        | 678 KB    | : 100% 1.0/1 [00:00<00:00,  5.80it/s]\n",
            "_libgcc_mutex-0.1    | 3 KB      | : 100% 1.0/1 [00:00<00:00, 28.59it/s]\n",
            "cycler-0.10.0        | 9 KB      | : 100% 1.0/1 [00:00<00:00, 33.21it/s]\n",
            "setuptools-49.6.0    | 947 KB    | : 100% 1.0/1 [00:00<00:00,  2.82it/s]\n",
            "libgomp-9.3.0        | 376 KB    | : 100% 1.0/1 [00:00<00:00,  9.28it/s]\n",
            "bzip2-1.0.8          | 484 KB    | : 100% 1.0/1 [00:00<00:00,  5.78it/s]\n",
            "samtools-1.12        | 386 KB    | : 100% 1.0/1 [00:00<00:00,  8.16it/s]\n",
            "libstdcxx-ng-9.3.0   | 4.0 MB    | : 100% 1.0/1 [00:00<00:00,  1.08it/s]\n",
            "matplotlib-base-3.4. | 7.2 MB    | : 100% 1.0/1 [00:01<00:00,  1.88s/it]\n",
            "libdeflate-1.7       | 67 KB     | : 100% 1.0/1 [00:00<00:00, 27.89it/s]\n",
            "lcms2-2.12           | 443 KB    | : 100% 1.0/1 [00:00<00:00,  8.00it/s]\n",
            "readline-8.1         | 295 KB    | : 100% 1.0/1 [00:00<00:00, 10.16it/s]\n",
            "ncurses-6.2          | 985 KB    | : 100% 1.0/1 [00:00<00:00,  1.72it/s]\n",
            "fontconfig-2.13.1    | 357 KB    | : 100% 1.0/1 [00:00<00:00,  8.72it/s]\n",
            "wheel-0.36.2         | 31 KB     | : 100% 1.0/1 [00:00<00:00, 22.96it/s]\n",
            "pandas-1.2.4         | 11.8 MB   | : 100% 1.0/1 [00:03<00:00,  3.84s/it]\n",
            "pcre-8.44            | 261 KB    | : 100% 1.0/1 [00:00<00:00, 10.57it/s]\n",
            "pytz-2021.1          | 239 KB    | : 100% 1.0/1 [00:00<00:00,  6.45it/s]\n",
            "tk-8.6.10            | 3.2 MB    | : 100% 1.0/1 [00:00<00:00,  1.20it/s]\n",
            "biopython-1.78       | 2.6 MB    | : 100% 1.0/1 [00:00<00:00,  1.03it/s]\n",
            "python-3.7.10        | 57.3 MB   | : 100% 1.0/1 [00:10<00:00, 10.64s/it]               \n",
            "matplotlib-3.4.2     | 7 KB      | : 100% 1.0/1 [00:00<00:00, 21.16it/s]\n",
            "pillow-8.1.2         | 688 KB    | : 100% 1.0/1 [00:00<00:00,  4.66it/s]\n",
            "c-ares-1.17.1        | 109 KB    | : 100% 1.0/1 [00:00<00:00, 17.22it/s]\n",
            "libssh2-1.9.0        | 226 KB    | : 100% 1.0/1 [00:00<00:00,  9.56it/s]\n",
            "openssl-1.1.1k       | 2.1 MB    | : 100% 1.0/1 [00:00<00:00,  2.14it/s]\n",
            "xorg-libxau-1.0.9    | 13 KB     | : 100% 1.0/1 [00:00<00:00, 32.65it/s]\n",
            "gettext-0.19.8.1     | 3.6 MB    | : 100% 1.0/1 [00:01<00:00,  1.23s/it]\n",
            "perl-5.32.0          | 16.8 MB   | : 100% 1.0/1 [00:05<00:00,  5.79s/it]\n",
            "pyqt-5.9.2           | 5.7 MB    | : 100% 1.0/1 [00:01<00:00,  1.86s/it]\n",
            "cython-0.29.23       | 2.2 MB    | : 100% 1.0/1 [00:00<00:00,  1.34it/s]\n",
            "gstreamer-1.14.0     | 3.2 MB    | : 100% 1.0/1 [00:00<00:00,  5.14it/s]\n",
            "libedit-3.1.20191231 | 121 KB    | : 100% 1.0/1 [00:00<00:00, 14.32it/s]\n",
            "pysam-0.16.0.1       | 2.6 MB    | : 100% 1.0/1 [00:00<00:00,  1.43it/s]\n",
            "openjpeg-2.4.0       | 525 KB    | : 100% 1.0/1 [00:00<00:00,  6.88it/s]\n",
            "libnghttp2-1.43.0    | 808 KB    | : 100% 1.0/1 [00:00<00:00,  4.18it/s]\n",
            "freetype-2.10.4      | 890 KB    | : 100% 1.0/1 [00:00<00:00,  3.73it/s]\n",
            "qt-5.9.7             | 68.5 MB   | : 100% 1.0/1 [00:03<00:00,  3.44s/it]               \n",
            "numpy-1.20.2         | 5.8 MB    | : 100% 1.0/1 [00:01<00:00,  1.76s/it]\n",
            "zstd-1.4.9           | 431 KB    | : 100% 1.0/1 [00:00<00:00,  7.52it/s]\n",
            "dbus-1.13.6          | 572 KB    | : 100% 1.0/1 [00:00<00:00,  5.83it/s]\n",
            "xz-5.2.5             | 343 KB    | : 100% 1.0/1 [00:00<00:00,  8.79it/s]\n",
            "libgcc-ng-9.3.0      | 7.8 MB    | : 100% 1.0/1 [00:01<00:00,  1.64s/it]               \n",
            "python-dateutil-2.8. | 220 KB    | : 100% 1.0/1 [00:00<00:00, 16.52it/s]\n",
            "xorg-libxdmcp-1.1.3  | 19 KB     | : 100% 1.0/1 [00:00<00:00, 39.34it/s]\n",
            "sqlite-3.35.5        | 1.4 MB    | : 100% 1.0/1 [00:00<00:00,  2.99it/s]\n",
            "expat-2.3.0          | 168 KB    | : 100% 1.0/1 [00:00<00:00, 13.65it/s]\n",
            "libev-4.33           | 104 KB    | : 100% 1.0/1 [00:00<00:00, 17.27it/s]\n",
            "Preparing transaction: - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\bdone\n",
            "#\n",
            "# To activate this environment, use\n",
            "#\n",
            "#     $ conda activate lab\n",
            "#\n",
            "# To deactivate an active environment, use\n",
            "#\n",
            "#     $ conda deactivate\n",
            "\n",
            "Activated conda environment `lab'!\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "eDfJ0fCzsTTb"
      },
      "source": [
        "설정이 끝났으면, 이제 콘솔을 열어서 터미널로 접속할 수 있는 통로를 열어줍니다."
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "J2LqGiP_N26s"
      },
      "source": [
        "## Google Drive에 시퀀싱 데이터 다운로드\n",
        "이제 실습 수업에서 필요한 시퀀싱 데이터를 다운로드 받아서 설치합시다. 드라이브 공간만 충분하면 데이터는 계속 저장돼 있기 때문에 한 번만 하시면 됩니다. Colab에서는 작업용 데이터를 Google Drive 폴더 아래의 `binfo-datapack1`에 저장합니다. 실습용 데이터는 `/content/drive/MyDrive/binfo1-datapack1`에 풀리게 됩니다. 첫 번째에는 wget이 접속오류를 내는 경우가 많으니, 접속에러가 발생하면 다시 한 번 실행해 보세요. 1G씩 끊어지지만 자동으로 이어받아집니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "kIN484e7NSxR",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "4f325f5c-f7b4-45e5-87f7-71fa8390859b"
      },
      "source": [
        "!wget -O - --no-check-certificate https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar | tar -C /content/drive/MyDrive -xf -"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-05-11 11:03:52--  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n",
            "Resolving hyeshik.qbio.io (hyeshik.qbio.io)... 147.47.203.129\n",
            "Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n",
            "WARNING: cannot verify hyeshik.qbio.io's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:\n",
            "  Unable to locally verify the issuer's authority.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 6424668160 (6.0G) [application/octet-stream]\n",
            "Saving to: ‘STDOUT’\n",
            "\n",
            "-                    16%[==>                 ]   1.01G  7.67MB/s    in 2m 48s  \n",
            "\n",
            "2021-05-11 11:06:42 (6.12 MB/s) - Connection closed at byte 1081147093. Retrying.\n",
            "\n",
            "--2021-05-11 11:06:43--  (try: 2)  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n",
            "Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n",
            "WARNING: cannot verify hyeshik.qbio.io's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:\n",
            "  Unable to locally verify the issuer's authority.\n",
            "HTTP request sent, awaiting response... 206 Partial Content\n",
            "Length: 6424668160 (6.0G), 5343521067 (5.0G) remaining [application/octet-stream]\n",
            "Saving to: ‘STDOUT’\n",
            "\n",
            "-                    33%[+++==>              ]   2.02G  6.79MB/s    in 2m 43s  \n",
            "\n",
            "2021-05-11 11:09:27 (6.32 MB/s) - Connection closed at byte 2164292988. Retrying.\n",
            "\n",
            "--2021-05-11 11:09:29--  (try: 3)  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n",
            "Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n",
            "WARNING: cannot verify hyeshik.qbio.io's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:\n",
            "  Unable to locally verify the issuer's authority.\n",
            "HTTP request sent, awaiting response... 206 Partial Content\n",
            "Length: 6424668160 (6.0G), 4260375172 (4.0G) remaining [application/octet-stream]\n",
            "Saving to: ‘STDOUT’\n",
            "\n",
            "-                    50%[++++++===>          ]   3.02G  5.39MB/s    in 2m 52s  \n",
            "\n",
            "2021-05-11 11:12:22 (6.00 MB/s) - Connection closed at byte 3244735523. Retrying.\n",
            "\n",
            "--2021-05-11 11:12:25--  (try: 4)  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n",
            "Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n",
            "WARNING: cannot verify hyeshik.qbio.io's certificate, issued by ‘CN=R3,O=Let's Encrypt,C=US’:\n",
            "  Unable to locally verify the issuer's authority.\n",
            "HTTP request sent, awaiting response... 206 Partial Content\n",
            "Length: 6424668160 (6.0G), 3179932637 (3.0G) remaining [application/octet-stream]\n",
            "Saving to: ‘STDOUT’\n",
            "\n",
            "-                    67%[++++++++++==>       ]   4.03G  8.35MB/s    in 3m 4s   \n",
            "\n",
            "2021-05-11 11:15:30 (5.61 MB/s) - Connection closed at byte 4326177482. Retrying.\n",
            "\n",
            "--2021-05-11 11:15:34--  (try: 5)  https://hyeshik.qbio.io/binfo/binfo1-datapack1.tar\n",
            "Connecting to hyeshik.qbio.io (hyeshik.qbio.io)|147.47.203.129|:443... connected.\n",
            "Unable to establish SSL connection.\n",
            "tar: Unexpected EOF in archive\n",
            "tar: Unexpected EOF in archive\n",
            "tar: Error is not recoverable: exiting now\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "lnlM-S70C-tY"
      },
      "source": [
        "받은 데이터의 MD5 체크섬을 확인합니다. 한참 걸립니다. 각 파일들의 원래 체크섬은 다음과 같습니다. 한 번 비교해 보세요.\n",
        "\n",
        "```\n",
        "140aaf30bcb9276cc716f8699f04ddd6  CLIP-35L33G.bam\n",
        "f1b3336ed7e2f97d562dcc71641251bd  CLIP-35L33G.bam.bai\n",
        "328883a73d507eafbf5b60bd6b906201  RNA-control.bam\n",
        "02073818e2f398a73c3b76e5169de1ca  RNA-control.bam.bai\n",
        "b09550d09d6c2a4ce27f0226f426fdb1  RNA-siLin28a.bam\n",
        "fef112c727244060ea62d3f2564a07f6  RNA-siLin28a.bam.bai\n",
        "28bbd0c47d725669340c784f1b772c01  RNA-siLuc.bam\n",
        "43590fdc4d81905c0432e0d1cb8cfd5b  RNA-siLuc.bam.bai\n",
        "5c08a9297307bc83259e658c4474f0cc  RPF-siLin28a.bam\n",
        "a1bb3e29be412dfd7fd8d16b1b1acc4c  RPF-siLin28a.bam.bai\n",
        "f2eebf50943024d0116c9cd3e744c707  RPF-siLuc.bam\n",
        "dc24f69e8f571fc8be30f28ce5b84fcd  RPF-siLuc.bam.bai\n",
        "```"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "g8lC_YJRBCPf",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "f747308c-fd55-45ad-9fcc-db2875ac7f94"
      },
      "source": [
        "!md5sum drive/MyDrive/binfo1-datapack1/*"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "140aaf30bcb9276cc716f8699f04ddd6  drive/MyDrive/binfo1-datapack1/CLIP-35L33G.bam\n",
            "f1b3336ed7e2f97d562dcc71641251bd  drive/MyDrive/binfo1-datapack1/CLIP-35L33G.bam.bai\n",
            "328883a73d507eafbf5b60bd6b906201  drive/MyDrive/binfo1-datapack1/RNA-control.bam\n",
            "02073818e2f398a73c3b76e5169de1ca  drive/MyDrive/binfo1-datapack1/RNA-control.bam.bai\n",
            "b09550d09d6c2a4ce27f0226f426fdb1  drive/MyDrive/binfo1-datapack1/RNA-siLin28a.bam\n",
            "fef112c727244060ea62d3f2564a07f6  drive/MyDrive/binfo1-datapack1/RNA-siLin28a.bam.bai\n",
            "28bbd0c47d725669340c784f1b772c01  drive/MyDrive/binfo1-datapack1/RNA-siLuc.bam\n",
            "43590fdc4d81905c0432e0d1cb8cfd5b  drive/MyDrive/binfo1-datapack1/RNA-siLuc.bam.bai\n",
            "5c08a9297307bc83259e658c4474f0cc  drive/MyDrive/binfo1-datapack1/RPF-siLin28a.bam\n",
            "a1bb3e29be412dfd7fd8d16b1b1acc4c  drive/MyDrive/binfo1-datapack1/RPF-siLin28a.bam.bai\n",
            "f2eebf50943024d0116c9cd3e744c707  drive/MyDrive/binfo1-datapack1/RPF-siLuc.bam\n",
            "dc24f69e8f571fc8be30f28ce5b84fcd  drive/MyDrive/binfo1-datapack1/RPF-siLuc.bam.bai\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "kiQ5y4dTD27M"
      },
      "source": [
        "그리고, GENCODE Annotation 데이터도 미리 받아서 저장해 둡니다. 에러가 나면 다시 한 번 실행해 봅니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Efdd3x8hDMh4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d4aebafe-ae5a-4e9f-9b7b-579da6633eac"
      },
      "source": [
        "!wget --no-check-certificate -O /content/drive/MyDrive/binfo1-datapack1/gencode.gtf.gz http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "--2021-05-11 11:27:26--  http://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M27/gencode.vM27.annotation.gtf.gz\n",
            "Resolving ftp.ebi.ac.uk (ftp.ebi.ac.uk)... 193.62.197.74\n",
            "Connecting to ftp.ebi.ac.uk (ftp.ebi.ac.uk)|193.62.197.74|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 28359457 (27M) [application/octet-stream]\n",
            "Saving to: ‘/content/drive/MyDrive/binfo1-datapack1/gencode.gtf.gz’\n",
            "\n",
            "/content/drive/MyDr 100%[===================>]  27.04M   813KB/s    in 35s     \n",
            "\n",
            "2021-05-11 11:28:02 (784 KB/s) - ‘/content/drive/MyDrive/binfo1-datapack1/gencode.gtf.gz’ saved [28359457/28359457]\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "O3rM9ds6EMhB"
      },
      "source": [
        "다루기 쉽도록 Annotation 압축을 풉니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "KNyuSM6PEQvX",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "6ba6281d-2324-48c5-9c1b-708fc033912e"
      },
      "source": [
        "!gunzip /content/drive/MyDrive/binfo1-datapack1/gencode.gtf.gz"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "gzip: /content/drive/MyDrive/binfo1-datapack1/gencode.gtf.gz: No such file or directory\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PVXzIs5FE3Q4",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "bfe8742d-b1d2-4b6f-fb22-d52557ac7e65"
      },
      "source": [
        "!ls -al drive/MyDrive/binfo*"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "total 7136706\n",
            "-rw------- 1 root root 1370036258 May  6 08:22 CLIP-35L33G.bam\n",
            "-rw------- 1 root root    3118336 May  6 08:23 CLIP-35L33G.bam.bai\n",
            "-rw------- 1 root root  883334756 May  5 14:04 gencode.gtf\n",
            "-rw------- 1 root root 1003658801 May  6 08:38 RNA-control.bam\n",
            "-rw------- 1 root root    2276104 May  6 08:39 RNA-control.bam.bai\n",
            "-rw------- 1 root root 1260991122 May  6 08:47 RNA-siLin28a.bam\n",
            "-rw------- 1 root root    2710744 May  6 08:47 RNA-siLin28a.bam.bai\n",
            "-rw------- 1 root root  981684502 May  6 08:59 RNA-siLuc.bam\n",
            "-rw------- 1 root root    2606104 May  6 08:59 RNA-siLuc.bam.bai\n",
            "-rw------- 1 root root  737352902 May  6 09:04 RPF-siLin28a.bam\n",
            "-rw------- 1 root root    2482768 May  6 09:05 RPF-siLin28a.bam.bai\n",
            "-rw------- 1 root root 1055123479 May  6 09:02 RPF-siLuc.bam\n",
            "-rw------- 1 root root    2605912 May  6 09:02 RPF-siLuc.bam.bai\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ecEA1PPQUlIK"
      },
      "source": [
        "실습에 필요한 프로그램들 몇 가지를 추가로 설치합니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "neteyWsWEKBy",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "b3376546-9d07-4e33-8a29-787056514667"
      },
      "source": [
        "!conda install -y subread"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Collecting package metadata (current_repodata.json): - \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\bdone\n",
            "Solving environment: \\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\b/ \b\b- \b\b\\ \b\b| \b\bdone\n",
            "\n",
            "\n",
            "==> WARNING: A newer version of conda exists. <==\n",
            "  current version: 4.9.2\n",
            "  latest version: 4.10.1\n",
            "\n",
            "Please update conda by running\n",
            "\n",
            "    $ conda update -n base -c defaults conda\n",
            "\n",
            "\n",
            "\n",
            "## Package Plan ##\n",
            "\n",
            "  environment location: /root/conda/envs/lab\n",
            "\n",
            "  added / updated specs:\n",
            "    - subread\n",
            "\n",
            "\n",
            "The following packages will be downloaded:\n",
            "\n",
            "    package                    |            build\n",
            "    ---------------------------|-----------------\n",
            "    subread-2.0.1              |       h5bf99c6_1        22.8 MB  bioconda\n",
            "    ------------------------------------------------------------\n",
            "                                           Total:        22.8 MB\n",
            "\n",
            "The following NEW packages will be INSTALLED:\n",
            "\n",
            "  subread            bioconda/linux-64::subread-2.0.1-h5bf99c6_1\n",
            "\n",
            "\n",
            "\n",
            "Downloading and Extracting Packages\n",
            "subread-2.0.1        | 22.8 MB   | : 100% 1.0/1 [00:05<00:00,  5.06s/it]               \n",
            "Preparing transaction: - \b\bdone\n",
            "Verifying transaction: | \b\b/ \b\bdone\n",
            "Executing transaction: \\ \b\bdone\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Oe_54MZrEi4y"
      },
      "source": [
        "## Transcript count table 마련하기\n",
        "\n",
        "제공된 alignment와 annotation 파일들을 이용해서 transcript별 read count를 구해봅니다. 실제 연구에서는 multi-mapping 등을 고려해야 하지만, 여기서는 단순화해서 모두 무시합니다. 데이터가 있는 곳으로 이동해서 작업합시다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "IANz6DmLUkA6",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "47c6f44b-69e6-498f-d09a-0d14aafc4a6e"
      },
      "source": [
        "%cd /content/drive/MyDrive/binfo1-datapack1/"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/content/drive/MyDrive/binfo1-datapack1\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "5OrnDxegFAip",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "84333022-8fa5-4420-f83b-7ee834f7c0a1"
      },
      "source": [
        "!featureCounts -a gencode.gtf -o read-counts.txt *.bam "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m    _____ _    _ ____  _____  ______          _____  \n",
            "       \u001b[44;37m =====      \u001b[0m\u001b[36m   / ____| |  | |  _ \\|  __ \\|  ____|   /\\   |  __ \\ \n",
            "       \u001b[44;37m   =====    \u001b[0m\u001b[36m  | (___ | |  | | |_) | |__) | |__     /  \\  | |  | |\n",
            "       \u001b[44;37m     ====   \u001b[0m\u001b[36m   \\___ \\| |  | |  _ <|  _  /|  __|   / /\\ \\ | |  | |\n",
            "       \u001b[44;37m       ==== \u001b[0m\u001b[36m   ____) | |__| | |_) | | \\ \\| |____ / ____ \\| |__| |\n",
            "       \u001b[44;37m ========== \u001b[0m\u001b[36m  |_____/ \\____/|____/|_|  \\_\\______/_/    \\_\\_____/\u001b[0m\n",
            "\t  v2.0.1\n",
            "\n",
            "//==========================\u001b[36m featureCounts setting \u001b[0m===========================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Input files : \u001b[36m6 BAM files  \u001b[0m \u001b[0m                                   ||\n",
            "||                           \u001b[32mo\u001b[36m CLIP-35L33G.bam\u001b[0m \u001b[0m                               ||\n",
            "||                           \u001b[32mo\u001b[36m RNA-control.bam\u001b[0m \u001b[0m                               ||\n",
            "||                           \u001b[32mo\u001b[36m RNA-siLin28a.bam\u001b[0m \u001b[0m                              ||\n",
            "||                           \u001b[32mo\u001b[36m RNA-siLuc.bam\u001b[0m \u001b[0m                                 ||\n",
            "||                           \u001b[32mo\u001b[36m RPF-siLin28a.bam\u001b[0m \u001b[0m                              ||\n",
            "||                           \u001b[32mo\u001b[36m RPF-siLuc.bam\u001b[0m \u001b[0m                                 ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||             Output file : \u001b[36mread-counts.txt\u001b[0m \u001b[0m                                 ||\n",
            "||                 Summary : \u001b[36mread-counts.txt.summary\u001b[0m \u001b[0m                         ||\n",
            "||              Annotation : \u001b[36mgencode.gtf (GTF)\u001b[0m \u001b[0m                               ||\n",
            "||      Dir for temp files : \u001b[36m./\u001b[0m \u001b[0m                                              ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "||                 Threads : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||                   Level : \u001b[36mmeta-feature level\u001b[0m \u001b[0m                              ||\n",
            "||              Paired-end : \u001b[36mno\u001b[0m \u001b[0m                                              ||\n",
            "||      Multimapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "|| Multi-overlapping reads : \u001b[36mnot counted\u001b[0m \u001b[0m                                     ||\n",
            "||   Min overlapping bases : \u001b[36m1\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n",
            "//=================================\u001b[36m Running \u001b[0m==================================\\\\\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Load annotation file gencode.gtf \u001b[0m... \u001b[0m                                      ||\n",
            "||    Features : \u001b[36m841952\u001b[0m \u001b[0m                                                      ||\n",
            "||    Meta-features : \u001b[36m55359\u001b[0m \u001b[0m                                                  ||\n",
            "||    Chromosomes/contigs : \u001b[36m22\u001b[0m \u001b[0m                                               ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file CLIP-35L33G.bam... \u001b[0m                                       ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m38880853\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m13630945 (35.1%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.92 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file RNA-control.bam... \u001b[0m                                       ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m24971058\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m5200836 (20.8%)\u001b[0m \u001b[0m                     ||\n",
            "||    Running time : \u001b[36m0.52 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file RNA-siLin28a.bam... \u001b[0m                                      ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m35108178\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m12338280 (35.1%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.71 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file RNA-siLuc.bam... \u001b[0m                                         ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m28117241\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m9748787 (34.7%)\u001b[0m \u001b[0m                     ||\n",
            "||    Running time : \u001b[36m0.76 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file RPF-siLin28a.bam... \u001b[0m                                      ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m30095551\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m10694162 (35.5%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.65 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Process BAM file RPF-siLuc.bam... \u001b[0m                                         ||\n",
            "||    Single-end reads are included. \u001b[0m                                         ||\n",
            "||    Total alignments : \u001b[36m42002435\u001b[0m \u001b[0m                                            ||\n",
            "||    Successfully assigned alignments : \u001b[36m12932285 (30.8%)\u001b[0m \u001b[0m                    ||\n",
            "||    Running time : \u001b[36m0.92 minutes\u001b[0m \u001b[0m                                            ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Write the final count table. \u001b[0m                                              ||\n",
            "|| Write the read assignment summary. \u001b[0m                                        ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "|| Summary of counting results can be found in file \"read-counts.txt.summary \u001b[0m ||\n",
            "|| \" \u001b[0m                                                                         ||\n",
            "||  \u001b[0m                                                                          ||\n",
            "\\\\============================================================================//\n",
            "\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "b2yJztZvGl9Q"
      },
      "source": [
        "결과를 pandas로 읽어와 봅니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BMyNT3ARFAWi",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272
        },
        "outputId": "04e5aa85-46ba-409e-b0ad-04dd41ecdfd9"
      },
      "source": [
        "import pandas as pd\n",
        "cnts = pd.read_csv('read-counts.txt', sep='\\t', comment='#', index_col=0)\n",
        "cnts.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chr</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Length</th>\n",
              "      <th>CLIP-35L33G.bam</th>\n",
              "      <th>RNA-control.bam</th>\n",
              "      <th>RNA-siLin28a.bam</th>\n",
              "      <th>RNA-siLuc.bam</th>\n",
              "      <th>RPF-siLin28a.bam</th>\n",
              "      <th>RPF-siLuc.bam</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Geneid</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000102693.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3143476</td>\n",
              "      <td>3144545</td>\n",
              "      <td>+</td>\n",
              "      <td>1070</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000064842.3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3172239</td>\n",
              "      <td>3172348</td>\n",
              "      <td>+</td>\n",
              "      <td>110</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000051951.6</th>\n",
              "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
              "      <td>3276124;3276746;3283662;3283832;3284705;349192...</td>\n",
              "      <td>3277540;3277540;3285855;3286567;3287191;349212...</td>\n",
              "      <td>-;-;-;-;-;-;-</td>\n",
              "      <td>6094</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000102851.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3322980</td>\n",
              "      <td>3323459</td>\n",
              "      <td>+</td>\n",
              "      <td>480</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000103377.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3435954</td>\n",
              "      <td>3438772</td>\n",
              "      <td>-</td>\n",
              "      <td>2819</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                     Chr  ... RPF-siLuc.bam\n",
              "Geneid                                                    ...              \n",
              "ENSMUSG00000102693.2                                chr1  ...             0\n",
              "ENSMUSG00000064842.3                                chr1  ...             0\n",
              "ENSMUSG00000051951.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1  ...             0\n",
              "ENSMUSG00000102851.2                                chr1  ...             0\n",
              "ENSMUSG00000103377.2                                chr1  ...             0\n",
              "\n",
              "[5 rows x 11 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 21
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "byKCv80ysiNr"
      },
      "source": [
        ""
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "NyrUZ7WBPAmv"
      },
      "source": [
        "이제 한 번 대충 그림을 그려 봅니다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "YlSZiJgVPAZS",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 272
        },
        "outputId": "6b50ab5e-eeb6-4ffc-bbbc-973126fc4612"
      },
      "source": [
        "cnts['clip_enrichment'] = cnts['CLIP-35L33G.bam'] / cnts['RNA-control.bam']\n",
        "cnts['rden_change'] = (cnts['RPF-siLin28a.bam'] / cnts['RNA-siLin28a.bam']) / (cnts['RPF-siLuc.bam'] / cnts['RNA-siLuc.bam'])\n",
        "cnts.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chr</th>\n",
              "      <th>Start</th>\n",
              "      <th>End</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Length</th>\n",
              "      <th>CLIP-35L33G.bam</th>\n",
              "      <th>RNA-control.bam</th>\n",
              "      <th>RNA-siLin28a.bam</th>\n",
              "      <th>RNA-siLuc.bam</th>\n",
              "      <th>RPF-siLin28a.bam</th>\n",
              "      <th>RPF-siLuc.bam</th>\n",
              "      <th>clip_enrichment</th>\n",
              "      <th>rden_change</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Geneid</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000102693.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3143476</td>\n",
              "      <td>3144545</td>\n",
              "      <td>+</td>\n",
              "      <td>1070</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000064842.3</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3172239</td>\n",
              "      <td>3172348</td>\n",
              "      <td>+</td>\n",
              "      <td>110</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000051951.6</th>\n",
              "      <td>chr1;chr1;chr1;chr1;chr1;chr1;chr1</td>\n",
              "      <td>3276124;3276746;3283662;3283832;3284705;349192...</td>\n",
              "      <td>3277540;3277540;3285855;3286567;3287191;349212...</td>\n",
              "      <td>-;-;-;-;-;-;-</td>\n",
              "      <td>6094</td>\n",
              "      <td>4</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>1</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>4.0</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000102851.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3322980</td>\n",
              "      <td>3323459</td>\n",
              "      <td>+</td>\n",
              "      <td>480</td>\n",
              "      <td>3</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>inf</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>ENSMUSG00000103377.2</th>\n",
              "      <td>chr1</td>\n",
              "      <td>3435954</td>\n",
              "      <td>3438772</td>\n",
              "      <td>-</td>\n",
              "      <td>2819</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "                                                     Chr  ... rden_change\n",
              "Geneid                                                    ...            \n",
              "ENSMUSG00000102693.2                                chr1  ...         NaN\n",
              "ENSMUSG00000064842.3                                chr1  ...         NaN\n",
              "ENSMUSG00000051951.6  chr1;chr1;chr1;chr1;chr1;chr1;chr1  ...         NaN\n",
              "ENSMUSG00000102851.2                                chr1  ...         NaN\n",
              "ENSMUSG00000103377.2                                chr1  ...         NaN\n",
              "\n",
              "[5 rows x 13 columns]"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 22
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "vU3ce_HEFALl",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 391
        },
        "outputId": "493ec554-b35a-4bd7-f3df-f19734f9a9dd"
      },
      "source": [
        "from matplotlib import pyplot as plt\n",
        "import numpy as np\n",
        "\n",
        "fig, ax = plt.subplots(1, 1, figsize=(5, 5))\n",
        "ax.scatter(np.log2(cnts['clip_enrichment']),\n",
        "           np.log2(cnts['rden_change']))"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/pandas/core/series.py:726: RuntimeWarning: divide by zero encountered in log2\n",
            "  result = getattr(ufunc, method)(*inputs, **kwargs)\n"
          ],
          "name": "stderr"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "<matplotlib.collections.PathCollection at 0x7f05fb37df90>"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 23
        },
        {
          "output_type": "display_data",
          "data": {
            "image/png": "iVBORw0KGgoAAAANSUhEUgAAATsAAAEvCAYAAAA6m2ZKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO2dfZBU53Xmn9M9d6AbO+rBGtuiDYKoHCgTBBPGFrXsbgXiFV6hjzFIwlppy5tslXZTcSpSlHGN1qoAjhKmMiuj1G42VVrHu6mSjcFCmSDjCrIDVaklC9LgGYTHhl3LQqCWvB4Hml0zLejpfvePnttzu/t9731v38/ue35VKjHdPd23p2+fez6ecw4JIcAwDNPtpKI+AIZhmDBgY8cwTCJgY8cwTCJgY8cwTCJgY8cwTCJgY8cwTCLoieJFb731VrFy5cooXpphmC7mzJkzPxdC9Mvui8TYrVy5EhMTE1G8NMMwXQwRva26j8NYhmESARs7hmESARs7hmESARs7hmESARs7hmESARs7hmESARs7hmESgS86OyLKAfgqgF8FIAD8lhDif/rx3EwyGZ8sYOzYBbxbLGFZLoPhbasxNJCP+rCYDsYvUfGfAfhbIcSDRNQLIOvT8zIJZHyygKdfPodSuQIAKBRLePrlcwDABo9pG89hLBHdAuCfA/hLABBC3BRCFL0+L5Ncxo5dqBs6k1K5grFjFyI6IqYb8CNntwrADID/RkSTRPRVIlrS/CAiepyIJohoYmZmxoeXZbqVd4slV7czjA5+GLseAL8G4C+EEAMArgMYaX6QEOIFIcSgEGKwv1/ap8swAIBluYyr2xlGBz+M3TsA3hFCnJ7/+SXUjB/DtMXwttXIGOmG2zJGGsPbVkd0REw34NnYCSF+CuAyEZln4m8A+KHX52WSy9BAHvt2rEM+lwEByOcy2LdjHRcnGE/4VY39XQBfn6/E/gTAb/r0vExCGRrIs3FjfMUXYyeEmAIw6MdzMQzDBAF3UDAMkwjY2DEMkwjY2DEMkwgi2UHBJAPub2XiBBs7JhC4v5WJGxzGMoHA/a1M3GBjxwQC97cycYONHRMI3N/KxA02dkwgcH8rEze4QMEEglmE4GosExfY2DGBwf2tTJxgY8fEGtbqMX7Bxo6JLazVY/yEjR0TW+y0eipjx54go4KNHRNb3Gr12BNk7GDpCRNb3Gr1uGuDsYONHRNb3Gr1uGuDsYONHRNb3O6i4K4Nxg7O2TGxxo1Wb3jb6oacHcBdG8wCbOyYFjq1osldG4wdJIQI/UUHBwfFxMRE6K/LONNc0QQAI01Y0tuDa6UyGxAm1hDRGSGEdPkXe3ZMA7KKZrkiUCyVAXSHnKNTPVfGG2zsmAZ0KpdOwl4vBG2IwtLisUGNH1yNZRrQrVwGIecwDVGhWILAgiEanyz49hphaPHCeB+Me9jYMQ3ItG0ygpBzhGGIwtDisbg5nrCxYxpo1rb1ZQ0YKWp4TFByjjAMURhaPBY3xxPO2TEtNGvbwso/LctlUJAYBD8NURhavDDeB+MeNnaMI2EN4QzDEIWhxWNxczxhnR0TK8YnC9hzZLoudelNE8pVASGANBEeuWs5nh1aF/FROsPV2GhgnR3TUdyYq9b/fbOycDGuCIEXT10CgNgYPJVR45H08YMLFEyskFUymzlw+nJIR2MPS0w6C9+MHRGliWiSiL7t13MyyUOnYlmJIPUigyUmnYWfnt3vAfiRj8/HJBCdimWTEiYyWGLSWfhi7IjoYwC2A/iqH8/HJBcdUXNVIBahIs/P6yz88uyeB/BFAFWnBzKMHVZRsx1xCBXdTlJmosWzsSOiewH8TAhxxuFxjxPRBBFNzMzMeH1ZposZGsjj5MhW2EWrcQgV3U5SZqLFD+nJZgD3E9E9ABYD+CUielEI8Zj1QUKIFwC8ANR0dj68LtPlqDoRzPviAEtMOgdfRcVE9OsA/kAIca/d41hUzOgwPlnA8LfOolxtPEeNNGHswfWYePsKDpy+jIoQHSU4ZoKDRcVMR2J6TNaOir6sgd33rcXE21fqAmMgnoJjt3DXRbBwuxjTkawaOQrZmZsmwpv77gn9eLwiG4efMdKcA3QJe3ZM6PjtpVifb7GRkho6oFVw7MdxhOFx2QmU2dj5A7eLMb7jdxtV8/OVyvYKJ/N1/DiOsFrCWKAcPGzsGN/xu41Kp1+2+fF+HUdYLWEsUA4eNnaM7/jtpbj9PVOu4sdxhOVxsUA5eNjYMb7jt5fi9vfSRL4dR1geFwuUg4cLFIzv2E3qfWb8nFIbJysEAMD1G3OuXt8sUmxZ098gTzHZsqa/4fUKxRLSRKgIgbzldc37CGgoiATlcbFAOVjY2DG+oxp9bqeNG7x9aYOAuFAs4fcPTSFN1CIqzhopLDLSuDpblr6+2Vd74ry8LfHE+ZkWqYdpIAvFEoZfOgsI1F9XAHWDl2f9W8fCOjsmNO54+jvSWXRpInxwcU9dOOxEPpfByZGtUm0asCA8fvLglFSiQrBvRXN6XSa+sM6OqaMKFf3Qkan0aObtqqGbFSG0DR2wUByQdVgAwNXZMp5++RwyRgqzEplKLmu4NnTW12U6EzZ2CaLZEyoUSxj+1lmAgHJlIYx7+uVzAODK4Mme++mXz2Hi7Ss4fKZgKx0hAtwEGNbiwNBAHmPHLrQYy1K5opyacmP+PrcxDctAOhuuxiYImWasXBV1Q2fSjo5MpUc7cPqyo0bOzeBhWXFA5XGpjNlsuWpr6Iw0hbYYnAkPNnYJwg99mdvH2+2LSBMha6RQVTzESBMe27TCUY7hp8eVz2Uw9uB6jD20nmUgXQaHsQnCTVLerQFRPbcp6WjGTPavGjmqfM5dn6zJUqy5wD1HprH3lWkUZ8tYlstgy5p+19IUO6z5ynaNG08viSfs2SUImUrfSBGMtPeQTdUB8Mhdy1tCQqCW09s8ehy5rKF8zsNnCnhm/FxDb2qxVMbV2XK9T/XFU5dcFTec8Nr3yusV4wtLTxJGO9VYXU9F9dzDL51tyQuapAhIp0h5v8ozDJK+rIFsb09bntmGva9KjS/LVsLBTnrCxo6xZXyy0GKszEnBOgZg8+hxx9A5q5CIxAXduXLjkwU8cXBKeh8BeGt0ewBHx1hhnR3TNntfmW7xusoVgb2vTGt5dzo5wjgbOkB/rpxdBZtlK9HDxo6xRdWSJbtdprXzg4yRdjXiyUqK0FLtNVLUoC3UQac6bfcYlq1EDxcoEsL4ZAGbR49j1chRbB49HkjC3O3cOR1M2YcpA8llDPTZFDWaaTZ0BGDXp5Zj7MH19R5ac0qKHTqemeoxfVmDq7ExgD27BKDqbgCc5RW5jCFNuOcyrQYniHaq2Zs1WUlzcl+1g8IJgdoggMHblwKoGb+P3rLY1gvVrU6rpr3svm9tG0fK+A0XKBKAXZEglzFAhAbd2rfPvlc3cFkjhffnqg0ekpEijD20vqHv9d1iCSmNyqksrNTFOnFEp/BhRzvtYkDt7/EnO+5UXiTsRlgxwcMFioRj53FZvTZTt2ZltlxFimqhmGkQrQ3+sjFJdrRr6MzjMz1S1aw6Xdo9jNlyFb9/qFZxNQ2edS6e1YhWhMDhMwUM3r6Uw9gYwMYuAbQzzshKVQDZ3h5M/uHdDbcHkaNzIoj9DzLs9H1VUZu0IjNwzb8R9oYw7t5QwwWKBCDrbnCLzDv0q9rqlneLpUDHLRGAqoOXWiyV6+/fyUsMazQUd2/Yw8YuAVj3G7RLc6VxfLLgalqJDpvvWKpVaV2WywSqW/P7+cPS2IW1Ca1TYWOXEIYG8jg5shXP79rQlpdn9rKaXsLYsQtt571U/MObV5S6PivXb8xh5YeCMyCFYglXrt/wxZiHORqKd8/awzm7hNG8H+IWh2qsFWuBIIgvkK7xLJbKOPnmFd9f30qpXEUKtda4my7Ex0B0+ypUuVnu3qjBxi5BNCev9+/aIP0imlIJmbzDDItUX6x2JR1xpAqg/4OLAejnJ2Ub0zaPHg+lYGC31Y1hY5cY2hEW24VF+3dtaPliEYB/csdSXPzHUq1K6XLcehwpFEt4XvJejTRhSW9tSZBMbgIAR994ryEsb3fkvS6qrW5cja3Bxi4hOCWvZV8Qu7BoaCDfshpRAHjt4lXs+uRyx70TnYKZt9u3Y13DjtlyRWDJoh4QtfYJl8oVfP3UJamH61aK4lZKwrtn1bCx84mo9U1Or6/y0kxvQ+bxycIiwsKS6aNvvNfyfOWKwDdOX/IkHo4TArULgdmupjvowO7t6+Y7vbT5Ma14rsYS0XIiOkFEPySiaSL6PT8OrJOIWt+k8/qqJHWayEGu0Pi1FahNEB6fLCgrp91i6ExM4+SXiFq3YMBSEn/xQ3oyB+ApIcQnAGwC8DtE9AkfnrdjCOqk1J1UovP6qrHpqi6BBY+vddZcqVzBU4fOun07sSRjOH8FbpkfeuBHBdpNwcDOGw9qck0349nYCSHeE0J8f/7f/w/AjwAkyscOQt/kxlvUeX2rsNi6MUslNJZ5fFbCHpUeFDJj3ow5AcoPCcfOjfo5NbvXM3f+Dnz51UDHdnUTvubsiGglgAEAp/183rgThL5J5a2ZHpX1C6P7+qrktUyu0A3FBb8ozofrw9tW48mDU56kNSfOz2g/VpYztVKuinoqgfN5zvjWQUFEHwBwGMATQoj/K7n/cSKaIKKJmRn9D7wTUIWIbvRNzSGrKvldEaLFw9N9/WfGz+GOp7+DlSNHccfT38Ez47Uvx2JLKJfLGJ5by7oNgdrfbmgg71lD6MbbN71xneGiAOfznPBlnh0RGQC+DeCYEOIrTo/vxnl2XqqxzVU3wFmc27ytyun1nxk/Jx2JlE4RKpaKgvm6fVkDv3h/DuVuqzZ44LFNK3Di/IynAQjtbBlzM6g06Yt9Ap1nR0QE4C8B/EjH0HUrXvRNspBVwN7gNXsITq9/4PRl6e2Vamu1Fahpx4w0xX7zV5gcOH0Zzz283ja0tIOwUFxwczEMcrl5kvAjjN0M4F8D2EpEU/P/3ePD8yYGVWgjoN6PIABXSel2CgrlisCNOfbsTCpCtEyQsQsw+7JGw+PMv6RbaZLuiC5uDbPHj2rs/xBCkBDiTiHEhvn/vuPHwSUF1dU4n8vguYfXK090N18a3bxPM91SdfUTc4JMPpdRet7m7omTI1uRyxjKoZ66r7dvh/1o9zSR1m7bJMMjnmKA6spdnL0JALYFg1K5gj1HpqX3mUWPlSNH2WgFgF2xYbGRwpMHpzDw5VelE2SA2sVK17sbGsgrzwEC8NzDekvLkwwbu5CwEwgPDeSxc2PriXr9ZgXDL9WkJidHtipDpmKp3PKlser0GP8wK9gAkO1Vh5ZXZ8sQUO/dNfEazhKARzetYEOnAffGhoBOj6NKf1WuiHrjuF2iurmhX2fTF+OeA6cv18c3zd70rkVUaSdl8FQTb7CxCwG7di5zS5edB2aGS8PbVuOJg1PSxzQ39LOhC4aKEBifLGDvK9O+ze0ztZOAnsFj49YeHMaGgF071zPj5/CkwoCZmAWMoYG8ckeDU3sX4w8EYPils1rj493AguDgYWMXAqpqay5rKOeemaQIDXKC3fetddXQz/hLb08KZZdj2nXhXRHBwsYuBFTtXEI4jzBPpxrLEmYxw5SSpImwc6O6Umd9HOOdG3PBCaxZEBwsbOxCQDVx5JpCkmClXBF46tDZesVufLKAg69drntyFSFw8LXL2LKmX2pQn3t4PS6Obseb++6xFcB63SvLeMMqCNYd7cW4w5feWLd0Y29sO9g1/DeTMdLYt2Md9hyZluq2chkD966/DQdO1wxhmgibfrmvvg/Cjr6sge133ibtnWX8ITe/xU2W60sT1XVysj5p87M3CxNRT8WOM3a9sezZRYhuGxCwkMBWCVSLpTIOnyk0eHwn37yiZUyvlco4+Jq8d5ZpJZexX+SdotYv1o25Kj5x2welj3/kruUNshK7QaxRT8XuZNjYRUhzeOuEUwK73WpsVYCnm7hAdcExqYraGkYrpXIFp35yVfp4q8bSaRArj2pvHzZ2EWP2Wb41ul0pKzFZlss4PoaJL6qKudXAqYoU5u12o9oZe1hUHAFmzsVcy1cRAvlcBr94395j4BO6OxGozaxblstgy5p+HHz9coO8xUhTvXhht5x8fLLAuTsb2LMLmeaeVfNqXyiWwGPjkouZfzv4+mVUmnV8lh+Ht62WpjzMlY+MGjZ2IePXOj6mOylXREu+r1wVdUNmNxqeRcn2sLELGT4hu58lNtNQ2sV63qgE5CxKtoeNXcjwCdn9XPdhGkoz1vPGjwVPSYSNXci40dYxySOl0CBtWdNf/7eqI4eLE/ZwNTZkzBPyqUNntZv3U1TTbjHdTT6XwezNOWmXRfO8Qx715B727CJgaCBvu1uiGTZ03U9f1sDJka31hdzNuBnhzshJpGcXh95Ca3sQ6+eYq7NlPDN+znbCtO6AT0ZO4jy7OPUWWrdUMcyLpy7Zpja4LcwbiTN2cewt9Ltokc9l2IB2KZ0gXYrriKrEhbFOjdZRYIYle1+Z9mXc95Y1/Ri8fWnbm+uZ+BKFdMlN2kdnuVRUJM6zc2q0NvHz6qTzXO2cCKrpwyfOz7Rsrmc6nyi0dG7TPnGMnEwSZ+x0BJmyD/jJg1MNO0N1cXOyuPHqCOopGoViCZtHjwOo7Zu9OLqdjV4HYqQIfVkjUi2dW+MVx8jJJHFhrM7uTdkHLAB8/dQlDN6+1NUJ57RGsV2c1CiFYglPHJzCniPT2HP/WgxvW81hbcwxUoQPLO5BcbYcmwnEbo2XaipLHDqHEmfsAGdBpuqDNCdLuDkB3ZwsuYzhOBjSLcVSGU8cnMKS3jRK5Up9pBQTP8pVgWxvD3bftzZyI2fi1njJLqpxaWVLpLFzQvUBA+7d8VzWkIan1pPFTAD7beismP2aFSFAcPYMmWiwJvQB+wgkDNwaL53IKSrY2EkY3rYaTx6ckhoEN+74+GQBv3h/ruV26zBG2YKVoGFDF29K5Qr2HJnGjblq5FXNdoxXXFvZEmvs7MrpQwN5TLx9pWWBtVt3fOzYBeluhyW9PfVNUm56ZJnkIPPy/cj1tkNcjZdbfDF2RPQZAH8GIA3gq0KIUT+eNyh0tEDPDq3D4O1LPbnjqpD3WqlcPwY2dIwb3KRR4tAWGSc8GzsiSgP4cwD/AsA7AF4noiNCiB96fe6g0K2Q2l3R7E4k8z6VGVuWy/DEYqYtdNIo45OFFoF6nMS9UeGHZ/cpAD8WQvwEAIjomwAeABBbY+dVC2TnGQKwzcEZKarnBBnGDea5Y4ddDthtGNxtnqEfxi4PwLph+R0Ad/nwvIHhVQvkJLS09djI/hgYRsUHFvc4GhuniMGPC3qnGrzQOiiI6HEimiCiiZmZGedfCBCvY63tPEOnk6lcEXjq0Fms/FD0Ikums1DNurPidP75dUHvRPwwdgUAyy0/f2z+tgaEEC8IIQaFEIP9/f3Nd4eK17HWdv21OidTRQicfPOK8n4C8PyuDTy+PWHkcxk8tmmFsudZ59yye4xfF/ROxQ9j9zqAjxPRKiLqBfA5AEd8eN5AMWfJvTW6HSdHtrpyze08Qz/GNS3LZTA0kMfOjfbHpFhXwMQQAvDYphW2jzk5shXPDq2TTrHWNVSq8y+XMXy7oHcqnnN2Qog5IvoCgGOoSU++JoSY9nxkMUa3v7ZQLLnuViCgflI37x1ohkUrnYP5WanOB0ItT2ZVADhV+837tqzpx4nzM/Wfd27MN/zcTmEhzm1f7UIiAp3X4OCgmJiY0Hpsp1eE2hEOP79rA49r70KcLnz5XAYnR7Y6ypqcOm4yRtqXCSmd+N0jojNCiEHpfXE2drIP1q8PMkzctITlMkZDmxCTHAjA/l0bWs4VI01Y0tuDa6Wy7Y4KK6bhTBp2xi7W8+y6qSJEGkEnASBykK4wXYtKbF6uCBRLZQioZxg208mFhKCItbHrhoqQ6dXNlqu2jyMAj25aoSUvYLqTLWv6fTu3O7mQEBSxNnbdUBHSbQvbv2sDnh1a11HvjfGXA6cvI2N4/0pai1zMArE2dl7Fv3FA90ptTkG5fqN1JBSTDCpCOEYAOgh0bpdDkMR6xFOcBwHqotMWliJg5chRHqrJaJEx0o7RwubR4x33XQmaWBs7oLNnael6aubIOzZ0ycHthS2fyzRc8J2kSd3Qy+o3sQ5jOwHVmkSzMNE8hDFrpOobo1RtQUz3I1Dz6HVIUa14sWze4I0du4Ata/odO3U6VbkQFLHW2cUdmX7OvGKrFttY9U+rRo6yN5dgjBQBVJOWuCVjpBs6JVTPQADeGt3ecFvcxcJejq9jdXZxR7VyEVDroawFC668dj+EeaMmoVwVWNLbg1zGcP28pXIFJ87P1Pu7VXuBZcvf3Sy9Dpsgj4+NXZuMTxbaaueynnx+DA1g4ktvmrB/1waMPbRe+ZhrHjbKWS+cKuXCljX9DWmWva9Mx1qoH2QjARu7NjCvPm5pls2Yo6aY7uRmRWDPkdpMDJXndYuHXcHWC6dsbNnOjXkcPlNo8JJkaz2B+Aj1g2wkiH01No642R9B80k8We7BzE0w3Yt1SbmRopZtc+22BhJqxssqMWlWLmwePa79/HFJqXidIm5HIoyd3wlZN1cZEsBXdm1oeb0o9sUy0XH9ZkVafb0x5ywizhop7Nj4MZw4P9MyNsxOYqJ7nsZJqB/kaKmuCGNV8g/zPr8Tnm6uMlWgHspY4e1iyUOyQliLviWL8OzQOpwc2Yp8LtNSeVXltFTnaS5jtD2lO2i8ThG3o+M9O6fFILprE92guvqojJcsJxOXHAkTf6znipucluo83XP/2tgYNxlBNRJ0vGfnVL0JIuHZfPXJZQwsdtnAHZccCRMPVAUMoPFccTMcI0gvqRPpeM/OyZj5mfCU5f4A+z2xANCXbdVRya66THIxJxQ75avc5rQ6ud3Sbzre2DkZM78SnqpweVFPytZgGWnC7vvWttw+NJDHxNtX8OKpS66Og+k+TK/ObvCF9UKbyxpY1JPCtVI5lh0QcaXjjZ2TMfNrcooqXLYzdHmH13JaqMN0PzLtpVPl/upsGRkjjf2SKj+jpuONnY4x88OVd5vjSxPVm7atx2mFF+okF4JceynDKS8d5z7XONHxxg4IJy+hM5fOitkby6N2mGbcLsNRXWjNc0ulRGAa6fhqbFh46WONU+8hEz1Xr99wpfNUFdPSRLHuc40bbOw0kZXx3UyrYF0dYzJbrmL4pbPaBk/V5K8zWYdZgI2dJjLZyZ7717achKp5jCmi2IzRYaKnPD8kQNX504y11YwA7NyY1x7rxNRgY6eBquUMQIu39+imFdJwtyJES5uanZCU6X6KpbJjG+P4ZAHDL53F9ZsL4aoAcPC1y9JpxXHqc40biZpU3O5AgM2jx6XFCVWieXyygKcOnXWcVMzDAJhmms8p1blnPtbcR8HV2Bp2k4q7ohqrg1MPrR1uW86GBvJ48uCU7e+YhrdUrtRHuKtGuTPJofmcssu/vVsscYeECxJj7LwMBFANWMwYKWwePS69qqqkKrdkDAx8+dWGIYqmgWNDxzTn2+wkT5ybc0dicnZeBgKoloDNlqvKnIusgmakCNdvzimnxTLJovm0kuXbhrethpFuPQGNFHFuziWJMXZupkU0U9Q0Ts0aJ+sklFzGwAcW97S1SYrpXPK5DDbfsbTl9oyRxqObVjhOJBkayGPswfUNwyRyGQNjD63n8NUlnsJYIhoDcB+AmwDeBPCbQoiiHwfmN14GArjpnni3WJIWHm7MVbkQkTCMFGHLmn4cPtNYYTWlI88O6e0f4bycP3j17L4L4FeFEHcC+F8AnvZ+SMHgZbaXLCRV6emW5TLK/CAvxU4W5arAi6cuSddtqoZA2E3dZrzhybMTQrxq+fEUgAe9HU6wtHuFlA0bMK/YMk9RVYnlAgRjIssVe1EMMM74prMjolcAHBRCvOj02Kh0dk641eGpHm+njWIYoNbXWhUCt2QMENXywimF9Mjt4IAk40lnR0TfA/BRyV1fEkL8zfxjvgRgDsDXbZ7ncQCPA8CKFSs0Djtc2rmqqjxFnkLMOGEaNaukiXtdg8XR2AkhPm13PxH9GwD3AvgNYeMmCiFeAPACUPPs3B1m8Pi5mMca9rKHx5gQoPTe7GA9nT94KlAQ0WcAfBHA/UKIWX8OKRrc6PB0kshDA3mcHNnqajIK071kjRTeGt2OqktDx72u/uG1g+I/A1gE4LtUqzSeEkL8e89HFQG6i3nswl2gtYhx/eZcsAfOxJ4UAX+y404AejImM5/Hva7+kqhBAHaomvL7sgZ237ewZ1NVfOjLGni/3Kils25ut5IiQIjaiX/l+g2Uys5b4ZnOQmWwnIY/ZIw09u2o6e/MC2fGSKE0V4UQted95K7l2ho98zWTMiyABwFoYH74e45MNySNr86WGzw31VVZ1gKmuowIAbw1ur0+vofpLgjAcw/LOxyaZUzWaqxqPees5WJYEaK+kU7H4LGcZQE2dhaGBvIYO3ahpem/VK5g7yvTeN8nD8wMjceOXeD2sS7k0U0rtA3JkkU9LZ7WwJdfdazkHzh9WcvY+Vl463TY2DWhKlTYNe9njDQW9aSkk1GaQ1lC7eqqq8XLKSauMPGkL2vYGiEnT2t8sqA1KEK3outlAEa3kZhBALq0U+bft2OddES7tdkbaDR8hWJJ2XJmZcminoYmcCa+qBaim5hDXZ3WIuqg23qoOp9vyRiJa0tjY9eEarmJSkKSz2Xq4mJZ7+2zQ+twcmQr8rlMSw5PQN1ja/JusQTuMusMyhWBibevSO8zPTon4bCux7Xpl/u0jJXdqDGnkfDdBoexTaiWbgNwnJpi13urOokFaqGPKnRZlsskMuToVL5+6hIGb1/ach7IcmdWTA9MR5ry8Q8vwfcvXdMqOsjO51nJTMUk5PHY2EmwM1oyI6iaVmxFdRL3ZQ1ke3ukxi5jpLFlTT8OnL7MQwQ6BAFIjYbdBct60dRpNfzJzGzL+WBnrJrP51UjR6XP2+0XVTZ2Lmg+adyU9WUnsZEm/OL9xqusmdfLWyarsKHrLFTidNntaaKGUbLuCKoAABFlSURBVGM6rYZee2h1BfTdBufsPGBX1m/GzOlZiw3likC52njimobu5MhWnDg/w8MEOpQNe19tyKepcsEyPZ7ZaqhatakqTugaK9WxdHtbGhs7D7RT1tfR6rlNVgO1k/WxTSv4Aw0JQi13pqJYKkt3DFsLXdax/TJURumRu5Z7MlZeBtl2MhzGesBtOLDnyLSWp+YmWQ0s7A8dGsjj6Bvv8UKfEBAA3rn6vtZjTW9/eNtq3JhbuNhZu3N0ui2sOeHB25d6agFL4qh37o31gKzP0extbD6RxicLeEIxwbiZ53dtqAtMdebiXRzdXv/3SkXymYke1V5gHs7pH3a9sRz1eMBNOKArFl3Sm25IVpvPb4c1N8R7LuKLqrBgdtR0u84tatizC4lVI0eVgwGsGCmSrsnTaS/LGGkuaHQwqqiA0Yc9uxigWykrV4XUC5Qlq5vhDWbxoZ3PQVXJZ/yBjV1I6BgrE1kVtjlkVlERQvt1mGAwJSWq9EPK5gPsdmFvlLCxCwlZfk/Vb6vyAk391f5dG5Seg5k3ZA/PG3YGqRkjRVjSu3CBWdRT+1qppCO/tFg92KHbhb1RwsYuRExj9dbodpwc2SqdlGIdASVLWNs1lJvtZWPHLnDXhQ2b71jq6P1WXfz5iICblQVJSbFUbtDWNRewrtmM7Op2YW+UsM4uQppbg5pHQMk0WKqG8jQRdm7MtyzuZlp57eJV7Prkct96jm9KBrCa+beTI1ulQwFUfdJcnAgO9uwixtoa1PyVkSWs7folub1Mj3KlNto8aO9X9Vmpwlu7WXiMd9iziwlOrWfjkwXsOTKt/H0z/GXixYa9r+JaqdzQ5WDXGcEEBxs7G8LcymTXeqbTSSGgVugz0WGO1G9OSySxXStqOIxVYBqYsKa5qkKbLWv6paO8ZbDsJN6wji5a2NgpcDO+yQ9k0hSz4KDrrZlFiiRAqFVV40LWYYKJCevoooPDWAVBbWVyExoffeM9VwWHihA4+sZ7no6vU8gYKTw0uAJTl6/h+s3gizIpAHbDuXp70hAgx8+LdXTRwZ6dAtVJ6eVktQuNZfe5HdWUz2Vc/05f1ujI0He2XMXwS2fx2V/Lw0iHIKB2eIlrpbLj0AYjRayjixA2dgqCmOZqFxo7LWSxQkDLF7zdYyvOlju246JcEThw+jJ2fXJ54MfvJDJeNr9lzm7C8AcW93BRIkLY2CkIYpqrXWisGx5njDT279qAsQfXS49N1YKmMgXml7TaoVXcihA4fKYgnd5rpCgUr4/Q2Pmg+iyLPFQ1UjhnZ4Pf8gCnycay+3IZA0sW9UhzfLJj23P/Wgx/62zDbosUABBa9s8a6YWwSncqspXHNq3A4O1L8eShqUh325bKFRw4fRmP3LUcJ87PNPytJt6+ghdPXVL+br6N993Mo5tWNHwWSV1oE3fYswsRu9BYdd+e+9c29NM6Gd+hgTzGHmr0+m7JGtIwrFwRePLgFDaPHseWNf1SLyiF1pAZqFVCnx1aBwBKbzJMTA9veNtqvDW6HcPbVmPs2AVbQ5cmwpY1/VrPL/MSCTWDb/4dTJK60Cbu8PDOkLGrxgYlYtYZHJox0ti5sXGHRS5jgAjSooe590JnbHyYuD0uaz+yijQRnnt4PQD9rocwBenMAnbDO9nYJQCdKceAfBeCylAS2gt9O40gpgezIQyOwCcVE9FTRCSI6FY/no9pZXyygM2jxxv2TeiiGz5ZE+vm66kuhctymUQIZHduzGPs2IW2/u4ywu7MYRbwbOyIaDmAuwGokyOMJ7x+QeyqtFbMBLr19VSs/FCm6xPufVkDh88UfDVMYXfmMAv44dntB/BFOKc+mDbx4wty7/rbbO+3JtB1NH//8OYV7eS+LmEo/XRfI2OkIQR8N0xBdeYwzngydkT0AICCEOKsxmMfJ6IJIpqYmZnx8rKJw48vyInz6r95s4ZQ53nF/HPq9oTqEPTVsi9raL0GoRa+qiYKezFMQXTmMHo46uyI6HsAPiq560sA/gNqIawjQogXALwA1AoULo4xUViT17msASHURsDNF0T1BSWgpSihW3goFEvhtGr5xDVNUa9pyJ3GbrVTZJBVilmWEg6Oxk4I8WnZ7US0DsAqAGep1qrzMQDfJ6JPCSF+6utRJoTmuXV2fa5uvyBuhK660o00EcqSkeRO5G1E1DqkU4QPLupBsVR2NcPPrpG/mUKxJM1zmmO3rH8f1Qh9GTy4Mzra7qAQQpwD8GHzZyK6CGBQCPFzH44rkej2x+bb+IK48SjsdmNYf7ddfd3wttV48uCU8v6skcKNOaE0YpWqwJJFPZjaXQsqdKU1biAsDN406csa2H3fWtscqs5nwoM7o4E7KGKETi7IDDvtxKwyiYrbXt+hgXy9E6DZ5KSo/YXcuUxtqYxdCF4qV6W9rlbeLZbq7zUIrZ/MzGZ7a438qs+pMH9MTDzxrTdWCLHSr+dKKjq5Mjsj0RwGF4olPHFwCntfmcbu+9a69ihUnqbZeuZ2BLzZ/gYAW9b0K1u5BIDDZwrYuTGv3ACWyxqhd2+YRs7uc9INZ5nwYc/OZ7yKf+28Gac8nco4XZ0tt6UP06066nh4aaIGT9KuOgwsNPdXhGiRi6hkIUFjXmjsPifWzMUXNnY+4of41xpq9mWNWn8q9EZM2Rmndr6EutXeqsQgNWN6Z+aFQCf0NH/H6tflMobjoulm0kTwOu7OeqExPycVrJmLJzziyUe8Jq4Bb8lrpzDY7ZdQtyprN6LKJEXwJey8fnOu/pq6ubqqEHh00wrbCSjNLOlNI5ftVVZMhwbyymXXrJmLJ2zsfCRqdbyTcXLK9+05Ml2vQJqVx3071tVlErdkDFy/OdcgN7F6PMMvnVVKUao+hZ3lisDYsQuuJpssy2XqY5hUOUArKQL++LPOzf+smessOIz1kajV8WZ4pdKHqb6E45MFDH/rbIPU4upsGcMv1RpjzHl6U7vvVk5IHhrIY0lve9fO5pDdiXeLJQwN5LFzY17r8WZb27ND6/DmvntwcXQ7nt+1QZp3yxopfOXhDdoSEr+nWTPBwZ6dj8ThSm8aHl2F//hkAU8dOiv1dkwvqjl8U32Z7fJoGSOFUrlV1isbK7Vh76stGjcr5sXjxPkZrfYvWTHEL3Eva+Y6BzZ2PhIndbzOl9AsqNiFdW5CcLs82o25VkNnXgiaDfO962/DN05dknY8WEfJ6x6b6nFsqJIFGzuf6aQvkE7HhpsQ3C6P1jwW3my2B9CiDTx8poB/tWkFvn32vRYPzxoq6xYpwi4Y8HDOeMI5uwTj5BlZvSgdnCQZVgRqxYInDk5JK9gnzs9gavfdLbm1YmlBM7hlTb9jzi7sNAIP54wvbOw6CC+CZRl2Hk9f1sDYg+vbymHZLYq2ohM+q+Q8e45M4/CZQkPOjlBbBBRlwYCHc8YXDmM7BFUr2J4j09hz/9q2vtCqgopXA+HHIh7TECt3sEoKGALAxX8stRQ8TMIIL6OWHzFq2LPrEFT5NWtY55agpBPW520Ha+jpNt+mMiphhZdRy48YNbxdrENwWocok3DEgWaP1Inm8VWy388YaSw2UsoVj7K/g910lHZGZqlQHS/r78Ih8O1iTPA4eQZxDZPciH+B2vsYO3bBcTTV7vvWulpEbff38dPLY6FxfOGcXYeg2woWR9mDrvgXQEOICSxIeVTvQfe9OslU3PYw29FJ8qMkwWFsBzE+WcDeV6ZbwjczTAJam+3jEEI5heAq2gnNVcZeJ5wmAPt3bYjdxYLRh8PYLmFoII/JP6xpz2RhUlxlD6oQPE1kG966Dc3tihA6RZNbMgZr5LoYNnYdyNBAvt6cbx3RHlfZg2zYZcZI47mH1+Ot0e1KA+S2gulk7M2/m2wIQMZIg8j/PbFMfGBj10XEVfbglLRXGUO3nQ+6xl51PEXFNjfdi4Xfom/GX7hA0UXEYeqKCrukvV8DFNysi5Qdj5dhnDLRN++jiBds7LqIOE1dcYsfFUyvxt7L7/sxpZoJFjZ2XUa3yB7akdB4NfZefj+u+VJmATZ2TOzwEhJ6Nfbt/r6bEJqJBi5QMLEjrhIaO/wqsjDBwZ4dEzs6MSTs5HxpUmBjx8QOtyFhXFrkuiVf2q1wGMvEDjchIU8GZnRhz46JHU4hodWTSxG1TDxmyQcjg40dE0tUIWFzpVY12j3O+T0mGjiMZToKnY1oAEs+mFbYs2NCQ6eQ4PQYHY/NSBFmb85h1chRrooydTx7dkT0u0R0noimiehP/TgopvvQKSToPMZpXFQuYwAEXJ0tc8GCacCTsSOiLQAeALBeCLEWwH/05aiYrkMlFH7q0Nn6lJA9R6YdxcRO46KWLOpBuSIvWDDJxmsY+9sARoUQNwBACPEz74fEdCOq8NMsMNiNTLf+rlOlthMFyUw4eDV2vwLgnxHRHwN4H8AfCCFe935YTLfhtAPC6Xet2Il3uUeVUeEYxhLR94joB5L/HkDNWC4FsAnAMIBDRCSdtE1EjxPRBBFNzMzM+PommPgjCz91cNtfyj2qjApHz04I8WnVfUT02wBeFrWtPa8RURXArQBarJkQ4gUALwC1hTttHzHTkTSHnzIxMAD0ZQ1ke3vabv3iHlVGhdcwdhzAFgAniOhXAPQC+Lnno2K6Emv4qVomvf3O23DivDfPn3tUGRlejd3XAHyNiH4A4CaAz4sodjMyHYfMA9uyph+HzxR4tDkTCLw3lokNm0ePS4sL7eyPZZIJ741lOgKWjTBBwsaOiQ1xXQXJdAds7JjYwLIRJkh4EAATG1g2wgQJGzsmVrBshAkKDmMZhkkEbOwYhkkEbOwYhkkEbOwYhkkEbOwYhkkEbOwYhkkEbOwYhkkEbOwYhkkEkUw9IaIZAG/bPORWJGsuXtLeL5C895y09wtE855vF0L0y+6IxNg5QUQTqjEt3UjS3i+QvPectPcLxO89cxjLMEwiYGPHMEwiiKuxeyHqAwiZpL1fIHnvOWnvF4jZe45lzo5hGMZv4urZMQzD+EqsjB0RPURE00RUJaLBpvueJqIfE9EFItoW1TEGBRHtIaICEU3N/3dP1McUBET0mfnP8MdENBL18YQBEV0konPzn2vXbZoioq8R0c/mtwyaty0lou8S0f+e/39flMcIxMzYAfgBgB0A/t56IxF9AsDnAKwF8BkA/4WI3K+Xjz/7hRAb5v/7TtQH4zfzn9mfA/iXAD4B4JH5zzYJbJn/XGMjxfCR/47a99LKCIC/E0J8HMDfzf8cKbEydkKIHwkhLkjuegDAN4UQN4QQbwH4MYBPhXt0jA98CsCPhRA/EULcBPBN1D5bpoMRQvw9gCtNNz8A4K/m//1XAIZCPSgJsTJ2NuQBXLb8/M78bd3GF4jojfmwIHK3PwCS8jk2IwC8SkRniOjxqA8mJD4ihHhv/t8/BfCRKA8GiGAHBRF9D8BHJXd9SQjxN2EfT5jYvXcAfwHgj1D7YvwRgOcA/FZ4R8cEyD8VQhSI6MMAvktE5+e9oUQghBBEFLnsI3RjJ4T4dBu/VgCw3PLzx+Zv6yh03zsR/VcA3w74cKKgKz5HtwghCvP//xkR/TVq4Xy3G7v/Q0S3CSHeI6LbAPws6gPqlDD2CIDPEdEiIloF4OMAXov4mHxl/oQw+SxqxZpu43UAHyeiVUTUi1rR6UjExxQoRLSEiD5o/hvA3ejOz7aZIwA+P//vzwOIPGqL1SpFIvosgP8EoB/AUSKaEkJsE0JME9EhAD8EMAfgd4QQlSiPNQD+lIg2oBbGXgTw76I9HP8RQswR0RcAHAOQBvA1IcR0xIcVNB8B8NdEBNS+b98QQvxttIfkL0R0AMCvA7iViN4BsBvAKIBDRPRvUZtw9HB0R1iDOygYhkkEnRLGMgzDeIKNHcMwiYCNHcMwiYCNHcMwiYCNHcMwiYCNHcMwiYCNHcMwiYCNHcMwieD/AzMgoloR1+bmAAAAAElFTkSuQmCC\n",
            "text/plain": [
              "<Figure size 360x360 with 1 Axes>"
            ]
          },
          "metadata": {
            "tags": [],
            "needs_background": "light"
          }
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "ksVjIL10mV_v"
      },
      "source": [
        "이 그림에는 문제가 많이 있습니다. 논문의 그림처럼 한 번 만들어 봅시다~"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "RZ-EV3s5X3u5"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "sf6EJJ60Qbxg"
      },
      "source": [
        "## Protein localization 반영하기\n",
        "\n",
        "논문 그림에서는 protein localization을 반영해서 색으로 나타냈습니다. 이것을 한 번 구현해 봅시다. 각 gene product의 localization을 따로 구하려면 좀 귀찮기 때문에 미리 모아 두었습니다. UniProt에서 매핑된 데이터를 토대로 아주 명확한 것만 일부 정리한 것입니다. 이것도 마찬가지로 처음엔 에러가 나기도 하니까, 다시 실행해보면 거의 잘 됩니다~"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ivb1q7_f-gl4",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 204
        },
        "outputId": "6f3c345c-6057-42b5-b0ce-b2203e4ea31c"
      },
      "source": [
        "import ssl\n",
        "ssl._create_default_https_context = ssl._create_unverified_context\n",
        "mouselocal = pd.read_csv('https://hyeshik.qbio.io/binfo/mouselocalization-20210507.txt', sep='\\t')\n",
        "mouselocal.head()"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/html": [
              "<div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>gene_id</th>\n",
              "      <th>Gene names</th>\n",
              "      <th>type</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>ENSMUSG00000000001</td>\n",
              "      <td>Gnai3</td>\n",
              "      <td>cytoplasm</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>ENSMUSG00000000028</td>\n",
              "      <td>Cdc45 Cdc45l Cdc45l2</td>\n",
              "      <td>nucleus</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>ENSMUSG00000000049</td>\n",
              "      <td>Apoh B2gp1</td>\n",
              "      <td>cytoplasm</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>ENSMUSG00000000058</td>\n",
              "      <td>Cav2</td>\n",
              "      <td>cytoplasm</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>ENSMUSG00000000085</td>\n",
              "      <td>Scmh1</td>\n",
              "      <td>nucleus</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>"
            ],
            "text/plain": [
              "              gene_id            Gene names       type\n",
              "0  ENSMUSG00000000001                 Gnai3  cytoplasm\n",
              "1  ENSMUSG00000000028  Cdc45 Cdc45l Cdc45l2    nucleus\n",
              "2  ENSMUSG00000000049            Apoh B2gp1  cytoplasm\n",
              "3  ENSMUSG00000000058                  Cav2  cytoplasm\n",
              "4  ENSMUSG00000000085                 Scmh1    nucleus"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 25
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Q9vgSzeiQxDU"
      },
      "source": [
        "이제 이 localization 데이터와 위에서 만든 scatter를 결합해서 논문 그림과 비슷하게 만들어 봅시다."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "i_Zaez__VCXz"
      },
      "source": [
        ""
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}